diff --git a/vi.c b/vi.c
index 40178ca9..93e08abc 100644
--- a/vi.c
+++ b/vi.c
@@ -664,7 +664,7 @@ static int vi_motion(int vc, int *row, int *off)
 		break;
 	case '(':
 	case ')':
-		dir = mv == '(' ? 1 : -1;
+		dir = mv == ')' ? 1 : -1;
 		if (!bre)
 			bre = rset_smake("^[.?!]+['\\])]*(?:[ \t]+\n?|\n)", 0);
 		int subs[2], org;
@@ -704,8 +704,8 @@ static int vi_motion(int vc, int *row, int *off)
 	case '}':
 	case '[':
 	case ']':
-		dir = mv == '{' || mv == '[' ? 1 : -1;
-		mark = mv == '[' || mv == ']' ? '\n' : '{';
+		dir = mv == '}' || mv == ']' ? 1 : -1;
+		mark = mv == '[' || mv == ']' ? '{' : '\n';
 		for (i = 0; i < cnt; i++)
 			if (lbuf_sectionbeg(xb, dir, row, off, mark))
 				break;
