diff --git a/vi.c b/vi.c
index 0b2e9d9..48f020f 100644
--- a/vi.c
+++ b/vi.c
@@ -656,7 +656,7 @@ static int fs_searchback(int cnt, int *row, int *off)
 }
 
 /* read a motion */
-static int vi_motion(int *row, int *off)
+static int vi_motion(int *row, int *off, int comb)
 {
 	static char ca_dir;
 	static sbuf *savepath[10];
@@ -724,9 +724,13 @@ static int vi_motion(int *row, int *off)
 		break;
 	case 'w':
 	case 'W':
+		if (comb)
+			dir = 2;
+		else
+			dir = vi_nlword+1;
 		mark = mv == 'W';
 		for (i = 0; i < cnt; i++)
-			if (lbuf_wordbeg(xb, mark, vi_nlword+1, row, off))
+			if (lbuf_wordbeg(xb, mark, dir, row, off))
 				break;
 		break;
 	case '{':
@@ -1032,7 +1036,7 @@ static void vc_motion(int cmd)
 	o2 = o1;
 	if ((mv = vi_motionln(&r2, cmd)))
 		o2 = -1;
-	else if (!(mv = vi_motion(&r2, &o2)))
+	else if (!(mv = vi_motion(&r2, &o2, 1)))
 		return;
 	if (mv < 0)
 		return;
@@ -1327,7 +1331,7 @@ void vi(int init)
 			sbuf_cut(led_attsb, 0)
 		if (!vi_ybuf)
 			vi_ybuf = vi_yankbuf();
-		mv = vi_motion(&nrow, &noff);
+		mv = vi_motion(&nrow, &noff, 0);
 		if (mv > 0) {
 			if (strchr("|jk", mv)) {
 				noff = vi_col2off(xb, nrow, vi_col);
