diff --git a/vi.c b/vi.c
index 62da17b5..0e692895 100644
--- a/vi.c
+++ b/vi.c
@@ -664,9 +664,13 @@ static int vi_motion(int vc, int *row, int *off)
 		break;
 	case 'w':
 	case 'W':
+		if (vc && cnt == 1)
+			dir = 2;
+		else
+			dir = vi_nlword+1;
 		mark = mv == 'W';
 		for (i = 0; i < cnt; i++)
-			if (lbuf_wordbeg(xb, mark, vi_nlword+1, row, off))
+			if (lbuf_wordbeg(xb, mark, dir, row, off))
 				break;
 		break;
 	case '(':
