diff --git a/vi.c b/vi.c
index cc5df97..53cf93c 100644
--- a/vi.c
+++ b/vi.c
@@ -670,7 +670,7 @@ static void vc_status(int type)
 }
 
 /* read a motion */
-static int vi_motion(int *row, int *off)
+static int vi_motion(int *row, int *off, int comb)
 {
 	static char ca_dir;
 	static sbuf *savepath[10];
@@ -740,9 +740,13 @@ static int vi_motion(int *row, int *off)
 		break;
 	case 'w':
 	case 'W':
+		if (comb)
+			dir = 2;
+		else
+			dir = vi_nlword+1;
 		mark = mv == 'W';
 		for (i = 0; i < cnt; i++)
-			if (lbuf_wordbeg(xb, mark, vi_nlword+1, row, off))
+			if (lbuf_wordbeg(xb, mark, dir, row, off))
 				break;
 		break;
 	case '{':
@@ -1047,7 +1051,7 @@ static void vc_motion(int cmd)
 	o2 = o1;
 	if ((mv = vi_motionln(&r2, cmd, vi_arg ? vi_arg : 1)))
 		o2 = -1;
-	else if (!(mv = vi_motion(&r2, &o2)))
+	else if (!(mv = vi_motion(&r2, &o2, 1)))
 		return;
 	if (mv < 0)
 		return;
@@ -1311,7 +1315,7 @@ void vi(int init)
 			sbuf_cut(led_attsb, 0)
 		if (!vi_ybuf)
 			vi_ybuf = vi_yankbuf();
-		mv = vi_motion(&nrow, &noff);
+		mv = vi_motion(&nrow, &noff, 0);
 		if (mv > 0) {
 			if (strchr("|jk", mv)) {
 				noff = vi_col2off(xb, nrow, vi_col);
