diff --git a/led.c b/led.c
index 3262a9d..777c637 100644
--- a/led.c
+++ b/led.c
@@ -286,7 +286,7 @@ static int led_lastword(char *s)
 	return r - s;
 }
 
-static void led_printparts(char *ai, char *pref, char *main, char *post)
+static void led_printparts(char *ai, char *pref, char *main, char *post, int print)
 {
 	if (!xled)
 		return;
@@ -306,8 +306,10 @@ static void led_printparts(char *ai, char *pref, char *main, char *post)
 		xleft = pos - xcols / 2;
 	if (pos < xleft)
 		xleft = pos < xcols ? 0 : pos - xcols / 2;
-	syn_blockhl = 0;
-	led_print(ln->s, -1);
+	if (print) {
+		syn_blockhl = 0;
+		led_print(ln->s, -1);
+	}
 	/* cursor position for inserting the next character */
 	if (*pref || *main || *ai) {
 		if (off - 2 >= 0)
@@ -384,7 +386,8 @@ static char *led_line(char *pref, char *post, char *ai,
 	if (!post)
 		post = "";
 	while (1) {
-		led_printparts(ai, pref, sb->s, post);
+		led_printparts(ai, pref, sb->s, post, !strchr("ABCD", vi_insmov));
+		vi_insmov = -1;
 		len = sb->s_n;
 		c = term_read();
 		switch (c) {
@@ -563,6 +566,20 @@ static char *led_line(char *pref, char *post, char *ai,
 		case TK_CTL('l'):
 			term_clean();
 			continue;
+		case '\033':;	/* Arrow keys */
+			char cbuf[1];
+			cbuf[0] = '\0';
+			int fl = fcntl(STDIN_FILENO, F_GETFL);
+			fcntl(STDIN_FILENO, F_SETFL, fl | O_NONBLOCK);
+			read(STDIN_FILENO, cbuf, 1);
+			if (*cbuf == '[') {
+				read(STDIN_FILENO, cbuf, 1);
+				vi_insmov = *cbuf;
+				fcntl(STDIN_FILENO, F_SETFL, fl);
+				goto _leave;
+			}
+			fcntl(STDIN_FILENO, F_SETFL, fl);
+			goto leave;
 		case TK_CTL('o'):;
 			preserve(int, xvis, 0)
 			term_exec(":", 1, /*nop*/, /*nop*/)
@@ -580,6 +597,7 @@ static char *led_line(char *pref, char *post, char *ai,
 	}
 leave:
 	vi_insmov = c;
+_leave:
 	*key = c;
 	sbufn_done(sb)
 }
@@ -639,7 +657,7 @@ sbuf *led_input(char *pref, char *post, int *kmap, int row)
 			return sb;
 		}
 		led_printparts(ai, pref ? pref : "", uc_lastline(ln),
-				key == '\n' ? "" : post);
+				key == '\n' ? "" : post, 1);
 		if (key == '\n')
 			term_chr('\n');
 		if (!pref || !pref[0]) {	/* updating autoindent */
diff --git a/vi.c b/vi.c
index 9fee3a7..003ac68 100644
--- a/vi.c
+++ b/vi.c
@@ -1737,6 +1737,29 @@ void vi(int init)
 				vc_insert(c);
 				vi_mod = !xpac && xrow == orow ? 2 : 1;
 				ins:
+				switch (vi_insmov) {
+				case 'A':	/* ↑ */
+					vi_back(c == 'a' || c == 'A' ? c : 'i');
+					xrow--;
+					xrow = xrow < 0 ? 0 : xrow;
+					xoff = vi_col2off(xb, xrow, vi_col);
+					goto _break;
+				case 'B':	/* ↓ */
+					vi_back(c == 'a' || c == 'A' ? c : 'i');
+					xrow++;
+					xoff = vi_col2off(xb, xrow, vi_col);
+					goto _break;
+				case 'D':	/* ← */
+					vi_back('i');
+					xoff--;
+					xoff = xoff < 0 ? 0 : xoff;
+					vi_col = vi_off2col(xb, xrow, xoff);
+					goto _break;
+				case 'C':	/* → */
+					vi_back('a');
+					vi_col = vi_off2col(xb, xrow, xoff);
+					goto _break;
+				}
 				if (vi_insmov == 127) {
 					if (xrow && !(xoff > 0 && lbuf_eol(xb, xrow))) {
 						xoff = lbuf_eol(xb, --xrow);
@@ -1750,6 +1773,9 @@ void vi(int init)
 					xoff--;
 				xoff = xoff < 0 ? 0 : xoff;
 				break;
+				_break:
+				vi_mod = 0;
+				break;
 			case 'J':
 				vc_join(1, vi_arg1 <= 1 ? 2 : vi_arg1);
 				break;
