diff --git a/led.c b/led.c
index 3262a9d..1903a92 100644
--- a/led.c
+++ b/led.c
@@ -2,6 +2,7 @@
 
 static sbuf *suggestsb;
 static sbuf *acsb;
+static int led_reins;
 
 int dstrlen(const char *s, char delim)
 {
@@ -288,7 +289,7 @@ static int led_lastword(char *s)
 
 static void led_printparts(char *ai, char *pref, char *main, char *post)
 {
-	if (!xled)
+	if (!xled && !led_reins)
 		return;
 	sbuf *ln;
 	int off, pos;
@@ -386,6 +387,10 @@ static char *led_line(char *pref, char *post, char *ai,
 	while (1) {
 		led_printparts(ai, pref, sb->s, post);
 		len = sb->s_n;
+		if (led_reins) {
+			xled = led_reins;
+			led_reins = 0;
+		}
 		c = term_read();
 		switch (c) {
 		case TK_CTL('h'):
@@ -563,6 +568,51 @@ static char *led_line(char *pref, char *post, char *ai,
 		case TK_CTL('l'):
 			term_clean();
 			continue;
+		case '\033':;	/* Arrow keys */
+			char cbuf[1];
+			cbuf[0] = '\0';
+			int fl = fcntl(STDIN_FILENO, F_GETFL);
+			fcntl(STDIN_FILENO, F_SETFL, fl | O_NONBLOCK);
+			read(STDIN_FILENO, cbuf, 1);
+			if (*cbuf == '[') {
+				read(STDIN_FILENO, cbuf, 1);
+				int tmp = xled;
+				xled = led_reins;
+				led_reins = tmp;
+				switch (*cbuf) {
+				case 'A':	/* ↑ */
+					if (!xoff)
+						term_push("ki", 2);
+					else
+						term_push("lki", 3);
+					if (xrow - xtop == 0) {
+						xled = led_reins;
+						led_reins = 0;
+					}
+					break;
+				case 'B':	/* ↓ */
+					if (!xoff)
+						term_push("ji", 2);
+					else
+						term_push("lji", 3);
+					if (xrow - xtop == xrows - 1) {
+						xled = led_reins;
+						led_reins = 0;
+					}
+					break;
+				case 'D':	/* ← */
+					term_push("i", 1);
+					break;
+				case 'C':	/* → */
+					if (!xoff)
+						term_push("li", 2);
+					else
+						term_push("lli", 3);
+					break;
+				}
+			}
+			fcntl(STDIN_FILENO, F_SETFL, fl);
+			goto leave;
 		case TK_CTL('o'):;
 			preserve(int, xvis, 0)
 			term_exec(":", 1, /*nop*/, /*nop*/)
