diff --git a/led.c b/led.c
index 3262a9d..9b22329 100644
--- a/led.c
+++ b/led.c
@@ -563,6 +563,27 @@ static char *led_line(char *pref, char *post, char *ai,
 		case TK_CTL('l'):
 			term_clean();
 			continue;
+		case '\033':	/* Arrow keys */
+			c = term_read();
+			if (c == '[') {
+				c = term_read();
+				switch (c) {
+				case 'A':	/* ↑ */
+					term_push("lki", 3);
+					break;
+				case 'B':	/* ↓ */
+					term_push("lji", 3);
+					break;
+				case 'D':	/* ← */
+					term_push("i", 1);
+					break;
+				case 'C':	/* → */
+					term_push("lli", 3);
+					break;
+				}
+			}
+			c = '\033';
+			goto leave;
 		case TK_CTL('o'):;
 			preserve(int, xvis, 0)
 			term_exec(":", 1, /*nop*/, /*nop*/)
