diff --git a/ex.c b/ex.c
index b4998fa..c77ec50 100644
--- a/ex.c
+++ b/ex.c
@@ -1123,7 +1123,7 @@ void ex(void)
 	}
 }
 
-void ex_init(char **files, int n)
+void ex_init(char **files, int n, char **cmds, int cmdnum)
 {
 	xbufsalloc = MAX(n, xbufsalloc);
 	ec_setbufsmax(NULL, NULL, "");
@@ -1134,4 +1134,6 @@ void ex_init(char **files, int n)
 	} while (--n > 0);
 	if (!(xvis & 2) && (s = getenv("EXINIT")))
 		ex_command(s)
+	for (int i = 0; i < cmdnum; i++)
+		ex_command(cmds[i])
 }
diff --git a/vi.c b/vi.c
index 702f8a5..9fcfaf2 100644
--- a/vi.c
+++ b/vi.c
@@ -2001,7 +2001,8 @@ static int setup_signals(void) {
 
 int main(int argc, char *argv[])
 {
-	int i, j;
+	int i, j, cmdnum = 0;
+	char *ex_cmds[argc - 1];
 	if (!setup_signals())
 		return EXIT_FAILURE;
 	dir_init();
@@ -2020,15 +2021,26 @@ int main(int argc, char *argv[])
 				xvis |= 4;
 			else if (argv[i][j] == 'v')
 				xvis &= ~4;
-			else {
+			else if (argv[i][j] == 'c') {
+				if (argv[i][j+1]) {
+					ex_cmds[cmdnum++] = argv[i] + j + 1;
+					break;
+				} else if (i + 1 < argc) {
+					ex_cmds[cmdnum++] = argv[++i];
+					break;
+				} else {
+					fprintf(stderr, "Missing argument for -c\n");
+					return EXIT_FAILURE;
+				}
+			} else {
 				fprintf(stderr, "Unknown option: -%c\n", argv[i][j]);
-				fprintf(stderr, "Usage: %s [-esv] [file ...]\n", argv[0]);
+				fprintf(stderr, "Usage: %s [-esv] [-c cmd] [file ...]\n", argv[0]);
 				return EXIT_FAILURE;
 			}
 		}
 	}
 	term_init();
-	ex_init(argv + i, argc - i);
+	ex_init(argv + i, argc - i, ex_cmds, cmdnum);
 	if (xvis & 4)
 		ex();
 	else
diff --git a/vi.h b/vi.h
index 95a93d5..760cddb 100644
--- a/vi.h
+++ b/vi.h
@@ -331,7 +331,7 @@ int ex_exec(const char *ln);
 char *ex_read(char *msg);
 void ex_print(char *line);
 void ex_show(char *msg);
-void ex_init(char **files, int n);
+void ex_init(char **files, int n, char** cmds, int cmdnum);
 void ex_bufpostfix(struct buf *p, int clear);
 int ex_krs(rset **krs, int *dir);
 void ex_krsset(char *kwd, int dir);
