diff --git a/ex.c b/ex.c
index 24ece75..8b1a3c0 100644
--- a/ex.c
+++ b/ex.c
@@ -1172,7 +1172,7 @@ void ex(void)
 	xgrec--;
 }
 
-void ex_init(char **files, int n)
+void ex_init(char **files, int n, char **cmds, int cmdnum)
 {
 	xbufsalloc = MAX(n, xbufsalloc);
 	ec_setbufsmax(NULL, NULL, "");
@@ -1185,4 +1185,6 @@ void ex_init(char **files, int n)
 	xvis &= ~8;
 	if ((s = getenv("EXINIT")))
 		ex_command(s)
+	for (int i = 0; i < cmdnum; i++)
+		ex_command(cmds[i])
 }
diff --git a/vi.c b/vi.c
index d998bfe..4bd221c 100644
--- a/vi.c
+++ b/vi.c
@@ -2011,7 +2011,8 @@ static int setup_signals(void) {
 
 int main(int argc, char *argv[])
 {
-	int i, j;
+	int i, j, cmdnum = 0;
+	char *ex_cmds[argc - 1];
 	if (!setup_signals())
 		return EXIT_FAILURE;
 	dir_init();
@@ -2032,15 +2033,26 @@ int main(int argc, char *argv[])
 				xvis |= 8;
 			else if (argv[i][j] == 'v')
 				xvis &= ~4;
-			else {
+			else if (argv[i][j] == 'c') {
+				if (argv[i][j+1]) {
+					ex_cmds[cmdnum++] = argv[i] + j + 1;
+					break;
+				} else if (i + 1 < argc) {
+					ex_cmds[cmdnum++] = argv[++i];
+					break;
+				} else {
+					fprintf(stderr, "Missing argument for -c\n");
+					return EXIT_FAILURE;
+				}
+			} else {
 				fprintf(stderr, "Unknown option: -%c\n", argv[i][j]);
-				fprintf(stderr, "Usage: %s [-emsv] [file ...]\n", argv[0]);
+				fprintf(stderr, "Usage: %s [-emsv] [-c cmd] [file ...]\n", argv[0]);
 				return EXIT_FAILURE;
 			}
 		}
 	}
 	term_init();
-	ex_init(argv + i, argc - i);
+	ex_init(argv + i, argc - i, ex_cmds, cmdnum);
 	if (xvis & 4)
 		ex();
 	else
diff --git a/vi.h b/vi.h
index 78589d7..cf76f97 100644
--- a/vi.h
+++ b/vi.h
@@ -349,7 +349,7 @@ int ex_exec(const char *ln);
 char *ex_read(char *msg);
 void ex_cprint(char *line, int r, int c, int ln);
 #define ex_print(line) ex_cprint(line, -1, 0, 1)
-void ex_init(char **files, int n);
+void ex_init(char **files, int n, char** cmds, int cmdnum);
 void ex_bufpostfix(struct buf *p, int clear);
 int ex_krs(rset **krs, int *dir);
 void ex_krsset(char *kwd, int dir);
