diff --git a/conf.c b/conf.c
index 4e2c53c..3221671 100644
--- a/conf.c
+++ b/conf.c
@@ -49,6 +49,8 @@ struct highlight hls[] = {
 	{"/", NULL, {9}, {0}, 0, 1}, /* <-- optional, used by hlw if set */
 
 	{"c", NULL, {14 | SYN_BD}, {1}, 0, 2},
+	{"c", "(?:(^[0-9]+[ \t]+[0-9]+).+\\\\\n$)", {14, 12}, {-1, 0}},
+	{"c", "^[0-9]+[ \t]+[0-9]+", {12}},
 	{"c", "^.+\\\\\n$", {14}, {1}},
 	{"c", "(/\\*[^&&*/]*)|([^\"&&/*]*\\*/)", {4 | SYN_IT}, {0}, 2},
 	{"c", NULL, {0, 9 | SYN_BGMK(12), 9 | SYN_BGMK(12)}, {1, -1, -1}, 0, 3},
diff --git a/reindex.sh b/reindex.sh
index 0b047d8..a51e7f3 100755
--- a/reindex.sh
+++ b/reindex.sh
@@ -1,9 +1,11 @@
 for p in *.patch
 do
+	echo "PATCH: $p"
 	patch < $p
 	git diff > ./tmp
-	patch -R < $p
+	patch -R < $p &>/dev/null
 	cp tmp $p
 	git add $p
+	echo "END PATCH: $p"
 done
 rm tmp *.orig
