diff --git a/ex.c b/ex.c
index d45b663..b516cbf 100644
--- a/ex.c
+++ b/ex.c
@@ -1126,3 +1126,11 @@ void ex_init(char **files, int n)
 	if ((s = getenv("EXINIT")))
 		ex_command(s)
 }
+
+void ex_done(void)
+{
+	for (int i = 0; i < xbufcur; i++)
+		bufs_free(i);
+	rset_free(xkwdrs);
+	free(bufs);
+}
diff --git a/regex.c b/regex.c
index 51d54a4..b45cbf0 100644
--- a/regex.c
+++ b/regex.c
@@ -540,6 +540,7 @@ int re_pikevm(rcode *prog, const char *s, const char **subp, int nsubp, int flg)
 	int *pcs[prog->splits];
 	rsub *subs[prog->splits];
 	unsigned int sdense[prog->sparsesz], sparsesz = 0;
+	memset(sdense, 0, sizeof(int) * prog->sparsesz);
 	rsub *nsub, *s1, *matched = NULL, *freesub = NULL;
 	rthread _clist[prog->len], _nlist[prog->len];
 	rthread *clist = _clist, *nlist = _nlist, *tmp;
diff --git a/ren.c b/ren.c
index f3b49de..cf5fd0d 100644
--- a/ren.c
+++ b/ren.c
@@ -78,6 +78,13 @@ void dir_init(void)
 	dir_rsctx = rset_make(i, ctx, 0);
 }
 
+void dir_done(void)
+{
+	rset_free(dir_rslr);
+	rset_free(dir_rsrl);
+	rset_free(dir_rsctx);
+}
+
 static ren_state rstates[1];
 ren_state *rstate = &rstates[0];
 
@@ -382,3 +389,10 @@ void syn_init(void)
 		pats[i] = fts[i].pat;
 	syn_ftrs = rset_make(i, pats, 0);
 }
+
+void syn_done(void)
+{
+	for (; ftmidx >= 0; ftmidx--)
+		rset_free(ftmap[ftmidx].rs);
+	rset_free(syn_ftrs);
+}
diff --git a/vi.c b/vi.c
index 88ce5c6..0da1955 100644
--- a/vi.c
+++ b/vi.c
@@ -600,6 +600,12 @@ void vi_regput(int c, const char *s, int ln)
 	vi_regputraw(c, s, ln, isupper(c));
 }
 
+static void vi_regdone(void)
+{
+	for (int i = 0; i < LEN(regs); i++)
+		free(regs[i]);
+}
+
 static void vi_regprint(void)
 {
 	term_pos(xrows, led_pos(vi_msg, 0));
@@ -2069,6 +2075,7 @@ int main(int argc, char *argv[])
 		ex();
 	else
 		vi(1);
+	ex_done();
 	term_done();
 	if (xvis & 4)
 		return EXIT_SUCCESS;
@@ -2077,5 +2084,12 @@ int main(int argc, char *argv[])
 		term_kill();
 	} else
 		term_clean();
+	vi_regdone();
+	syn_done();
+	dir_done();
+	led_done();
+	ren_done();
+	temp_done(1);
+	temp_done(0);
 	return EXIT_SUCCESS;
 }
diff --git a/vi.h b/vi.h
index 827274c..8e844b3 100644
--- a/vi.h
+++ b/vi.h
@@ -175,6 +175,7 @@ int ren_cwid(char *s, int pos);
 /* text direction */
 int dir_context(char *s);
 void dir_init(void);
+void dir_done(void);
 /* syntax highlighting */
 #define SYN_BD		0x010000
 #define SYN_IT		0x020000
@@ -196,6 +197,7 @@ int syn_merge(int old, int new);
 void syn_reloadft(void);
 int syn_addhl(char *reg, int func, int reload);
 void syn_init(void);
+void syn_done(void);
 
 /* uc.c utf-8 helper functions */
 extern unsigned char utf8_length[256];
@@ -333,6 +335,7 @@ void ex_print(char *line);
 void ex_show(char *msg);
 void ex_init(char **files, int n);
 void ex_bufpostfix(struct buf *p, int clear);
+void ex_done(void);
 int ex_krs(rset **krs, int *dir);
 void ex_krsset(char *kwd, int dir);
 int ex_edit(const char *path, int len);
