diff --git a/ex.c b/ex.c
index e41edeb..7fa73c6 100644
--- a/ex.c
+++ b/ex.c
@@ -1244,3 +1244,16 @@ void ex_init(char **files, int n)
 	if ((s = getenv("EXINIT")))
 		ex_command(s)
 }
+
+void ex_done(void)
+{
+	for (int i = 0; i < LEN(tempbufs); i++)
+		if (tempbufs[i].lb) {
+			free(tempbufs[i].path);
+			lbuf_free(tempbufs[i].lb);
+		}
+	for (int i = 0; i < xbufcur; i++)
+		bufs_free(i);
+	rset_free(xkwdrs);
+	free(bufs);
+}
diff --git a/regex.c b/regex.c
index 08470a5..0e95395 100644
--- a/regex.c
+++ b/regex.c
@@ -590,6 +590,7 @@ int re_pikevm(rcode *prog, const char *s, const char **subp, int nsubp, int flg)
 	int *pcs[prog->splits];
 	rsub *subs[prog->splits];
 	unsigned int sdense[prog->sparsesz], sparsesz = 0;
+	memset(sdense, 0, sizeof(int) * prog->sparsesz);
 	rsub *nsub, *s1, *matched = NULL, *freesub = NULL;
 	rthread _clist[prog->len], _nlist[prog->len];
 	rthread *clist = _clist, *nlist = _nlist, *tmp;
diff --git a/ren.c b/ren.c
index 305942a..5f0b449 100644
--- a/ren.c
+++ b/ren.c
@@ -93,6 +93,20 @@ static int ren_cwid(char *s, int pos)
 static ren_state rstates[2];
 ren_state *rstate = &rstates[0];
 
+void ren_done(void)
+{
+	rset_free(dir_rslr);
+	rset_free(dir_rsrl);
+	rset_free(dir_rsctx);
+	for (int i = 0; i < LEN(rstates); i++) {
+		if (rstate[i].col) {
+			free(rstate[i].col - 2);
+			free(rstate[i].pos);
+			free(rstate[i].chrs);
+		}
+	}
+}
+
 /* specify the screen position of the characters in s */
 ren_state *ren_position(char *s)
 {
@@ -378,3 +392,10 @@ void syn_init(void)
 		pats[i] = fts[i].pat;
 	syn_ftrs = rset_make(i, pats, 0);
 }
+
+void syn_done(void)
+{
+	for (; ftmidx >= 0; ftmidx--)
+		rset_free(ftmap[ftmidx].rs);
+	rset_free(syn_ftrs);
+}
diff --git a/vi.c b/vi.c
index ff4161d..c0ae9cb 100644
--- a/vi.c
+++ b/vi.c
@@ -525,6 +525,12 @@ void vi_regput(int c, const char *s, int ln)
 	vi_regputraw(c, s, ln, isupper(c));
 }
 
+static void vi_regdone(void)
+{
+	for (int i = 0; i < LEN(xregs); i++)
+		free(xregs[i]);
+}
+
 rset *fsincl;
 char *fs_exdir;
 static int fspos;
@@ -1924,7 +1930,10 @@ int main(int argc, char *argv[])
 		ex();
 	else
 		vi(1);
+	ex_done();
 	term_done();
+	free(ibuf);
+	rset_free(fsincl);
 	if (xvis & 4)
 		return EXIT_SUCCESS;
 	if (abs(xquit) == 2) {
@@ -1932,5 +1941,9 @@ int main(int argc, char *argv[])
 		term_kill();
 	} else
 		term_clean();
+	vi_regdone();
+	syn_done();
+	ren_done();
+	led_done();
 	return EXIT_SUCCESS;
 }
diff --git a/vi.h b/vi.h
index f00adb9..9ab3a98 100644
--- a/vi.h
+++ b/vi.h
@@ -220,6 +220,7 @@ char *ren_translate(char *s, char *ln);
 /* text direction */
 int dir_context(char *s);
 void dir_init(void);
+void dir_done(void);
 /* syntax highlighting */
 #define SYN_BD		0x010000
 #define SYN_IT		0x020000
@@ -242,6 +243,7 @@ void syn_reloadft(void);
 int syn_findhl(int id);
 void syn_addhl(char *reg, int id, int reload);
 void syn_init(void);
+void syn_done(void);
 
 /* uc.c utf-8 helper functions */
 extern unsigned char utf8_length[256];
@@ -401,6 +403,7 @@ void ex_cprint(char *line, int r, int c, int ln);
 #define ex_print(line) ex_cprint(line, -1, 0, 1)
 void ex_init(char **files, int n);
 void ex_bufpostfix(struct buf *p, int clear);
+void ex_done(void);
 int ex_krs(rset **krs, int *dir);
 void ex_krsset(char *kwd, int dir);
 int ex_edit(const char *path, int len);
