diff --git a/ex.c b/ex.c
index 750627a..5f10d90 100644
--- a/ex.c
+++ b/ex.c
@@ -963,6 +963,35 @@ static int ec_setbufsmax(char *loc, char *cmd, char *arg)
 	return 0;
 }
 
+static int ec_closebuf(char *loc, char *cmd, char *arg)
+{
+	int idx;
+	int istmp = istempbuf(ex_buf);
+	if (!*arg)
+		idx = istmp ? -1 : ex_buf - bufs;
+	else {
+		while (*arg == ' ' || *arg == '\t')
+			arg++;
+		idx = atoi(arg);
+	}
+	if (idx < 0 || (!idx && xbufcur < 2) || idx >= xbufcur)
+		return 1;
+	bufs_free(idx);
+	for (int i = idx; i < xbufcur; i++)
+		bufs[i] = bufs[i+1];
+	xbufcur--;
+	if (!istmp) {
+		ex_buf = idx < ex_buf - bufs ? ex_buf - 1 : ex_buf;
+		ex_buf = idx == ex_buf - bufs ? bufs : ex_buf;
+		ex_pbuf = idx == ex_buf - bufs ? bufs : ex_pbuf;
+	} else {
+		ex_pbuf = idx == ex_pbuf - bufs ? bufs : ex_pbuf;
+		ex_tpbuf = idx == ex_tpbuf - bufs ? bufs : ex_tpbuf;
+	}
+	syn_setft(ex_ft);
+	return 0;
+}
+
 static struct excmd {
 	char *name;
 	int (*ec)(char *loc, char *cmd, char *arg);
@@ -1009,6 +1038,7 @@ static struct excmd {
 	{"cd", ec_chdir},
 	{"inc", ec_setincl},
 	{"bx", ec_setbufsmax},
+	{"cx", ec_closebuf},
 	{"ac", ec_setacreg},
 	{"", ec_null},
 };
@@ -1097,3 +1127,11 @@ void ex_init(char **files, int n)
 	if (!(xvis & 2) && (s = getenv("EXINIT")))
 		ex_command(s)
 }
+
+void ex_done(void)
+{
+	for (int i = 0; i < xbufcur; i++)
+		bufs_free(i);
+	rset_free(xkwdrs);
+	free(bufs);
+}
