diff --git a/ex.c b/ex.c
index 57b642e..7154f99 100644
--- a/ex.c
+++ b/ex.c
@@ -946,6 +946,29 @@ static int ec_setbufsmax(char *loc, char *cmd, char *arg)
 	return 0;
 }
 
+static int ec_closebuf(char *loc, char *cmd, char *arg)
+{
+	int idx;
+	if (!*arg)
+		idx = ex_buf - bufs;
+	else {
+		while (*arg == ' ' || *arg == '\t')
+			arg++;
+		idx = atoi(arg);
+	}
+	if (idx >= xbufcur)
+		return 1;
+	bufs_free(idx);
+	for (int i = idx; i < xbufcur; i++)
+		bufs[i] = bufs[i+1];
+	xbufcur--;
+	ex_buf = idx < ex_buf - bufs ? ex_buf - 1 : ex_buf;
+	ex_buf = idx == ex_buf - bufs ? bufs : ex_buf;
+	ex_pbuf = idx == ex_buf - bufs ? bufs : ex_pbuf;
+	syn_setft(ex_filetype);
+	return 0;
+}
+
 static struct excmd {
 	char *name;
 	int (*ec)(char *loc, char *cmd, char *arg);
@@ -988,6 +1011,7 @@ static struct excmd {
 	{"cd", ec_chdir},
 	{"inc", ec_setincl},
 	{"bx", ec_setbufsmax},
+	{"cx", ec_closebuf},
 	{"ac", ec_setacreg},
 	{"", ec_null},
 };
