diff --git a/ex.c b/ex.c
index 5ba507c..746418a 100644
--- a/ex.c
+++ b/ex.c
@@ -1086,6 +1086,39 @@ static int ec_setenc(char *loc, char *cmd, char *arg)
 	return 0;
 }
 
+void ex_script(FILE *fp)
+{
+	char done = 0;
+	do {
+		size_t n = 128, i = 0;
+		int c;
+		char *ln = malloc(128);
+		while ((c = fgetc(fp)) != EOF && c != '\n') {
+			if (i >= n - 2) {
+				n += 128;
+				ln = erealloc(ln, n);
+			}
+			ln[i++] = c;
+		}
+		if (!i) {
+			free(ln);
+			done = 1;
+			break;
+		}
+		ln[i] = '\0';
+		ex_command(ln);
+		free(ln);
+	} while(!done);
+}
+
+static int ec_script(char *loc, char *cmd, char *arg)
+{
+	FILE *fp = popen(arg, "r");
+	ex_script(fp);
+	pclose(fp);
+	return 0;
+}
+
 static struct excmd {
 	char *name;
 	int (*ec)(char *loc, char *cmd, char *arg);
@@ -1138,6 +1171,7 @@ static struct excmd {
 	{"ac", ec_setacreg},
 	{"uc", ec_setenc},
 	{"ph", ec_setenc},
+	{"sc", ec_script},
 	{"", ec_print},
 };
 
