From e99f9ef368f69abb7013f010be24d2353fb496d0 Mon Sep 17 00:00:00 2001
From: OldWorldOrdr <joey.t.reinhart@gmail.com>
Date: Thu, 30 Nov 2023 14:32:21 -0500
Subject: [PATCH] support ~/.exrc

---
 ex.c | 34 ++++++++++++++++++++++++++++++++--
 vi.c |  1 +
 2 files changed, 33 insertions(+), 2 deletions(-)

diff --git a/ex.c b/ex.c
index aaac5be..002a945 100644
--- a/ex.c
+++ b/ex.c
@@ -1096,6 +1096,36 @@ void ex_init(char **files, int n)
 		ec_edit("", "e", s);
 		s = *(++files);
 	} while (--n > 0);
-	if (!(xvis & 2) && (s = getenv("EXINIT")))
-		ex_command(s)
+	if (!(xvis & 2)) {
+		if (getenv("EXINIT")) {
+			ex_command(getenv("EXINIT"));
+		} else if (getenv("HOME")) {
+			char exrc[PATH_MAX];
+			snprintf(exrc, sizeof(exrc), "%s/.exrc", getenv("HOME"));
+			struct stat st;
+			if (stat(exrc, &st) == 0) {
+				if (st.st_uid == getuid() && !(st.st_mode & S_IWGRP) && !(st.st_mode & S_IWOTH)) {
+					FILE *fp = fopen(exrc, "r");
+					if (fp) {
+						char *ln;
+						while (getline(&ln, &(size_t){0}, fp) >= 0) {
+							ln[strlen(ln) - 1] = '\0';
+							ex_command(ln);
+							free(ln);
+							ln = NULL;
+						}
+						fclose(fp);
+						if (ln)
+							free(ln);
+					} else {
+						fprintf(stderr, "Cannot open ~/.exrc\n");
+						exit(EXIT_FAILURE);
+					}
+				} else {
+					fprintf(stderr, "Bad permissions on ~/.exrc\n");
+					exit(EXIT_FAILURE);
+				}
+			}
+		}
+	}
 }
diff --git a/vi.c b/vi.c
index c6f6e9e..79dd7f9 100644
--- a/vi.c
+++ b/vi.c
@@ -18,6 +18,7 @@
  */
 #include <ctype.h>
 #include <fcntl.h>
+#include <limits.h>
 #include <stdio.h>
 #include <stdarg.h>
 #include <stdlib.h>
