diff --git a/ex.c b/ex.c
index 595f261..55d502a 100644
--- a/ex.c
+++ b/ex.c
@@ -21,6 +21,7 @@ int xgrp = 2;			/* regex search group */
 int xpac;			/* print autocomplete options */
 int xkwdcnt;			/* number of search kwd changes */
 int xbufcur;			/* number of active buffers */
+int xexrc;			/* read .exrc from the current directory */
 struct buf *bufs;		/* main buffers */
 struct buf tempbufs[2];		/* temporary buffers, for internal use */
 struct buf *ex_buf;		/* current buffer */
@@ -844,6 +845,7 @@ static struct option {
 	{"grp", &xgrp},
 	{"pac", &xpac},
 	{"led", &xled},
+	{"exrc", &xexrc},
 };
 
 static char *cutword(char *s, char *d)
@@ -1111,6 +1113,47 @@ void ex(void)
 	}
 }
 
+char *load_line(FILE *fp)
+{
+	char *ln = NULL;
+	int n = 0;
+	int c;
+	while ((c = fgetc(fp)) != EOF && c != '\n') {
+		ln = erealloc(ln, n + 2);
+		ln[n++] = c;
+	}
+	if (c == EOF && !n) {
+		free(ln);
+		return NULL;
+	}
+	ln[n] = '\0';
+	return ln;
+}
+
+void load_exrc(char *exrc)
+{
+	struct stat st;
+	if (stat(exrc, &st) == 0) {
+		if (st.st_uid == getuid() && !(st.st_mode & S_IWGRP) && !(st.st_mode & S_IWOTH)) {
+			FILE *fp = fopen(exrc, "r");
+			if (fp) {
+				char *ln = NULL;
+				while ((ln = load_line(fp))) {
+					ex_command(ln);
+					free(ln);
+				}
+				fclose(fp);
+			} else {
+				fprintf(stderr, "Cannot open ~/.exrc\n");
+				exit(EXIT_FAILURE);
+			}
+		} else {
+			fprintf(stderr, "Bad permissions on ~/.exrc\n");
+			exit(EXIT_FAILURE);
+		}
+	}
+}
+
 void ex_init(char **files, int n)
 {
 	xbufsalloc = MAX(n, xbufsalloc);
@@ -1120,6 +1163,16 @@ void ex_init(char **files, int n)
 		ec_edit("", "e", s);
 		s = *(++files);
 	} while (--n > 0);
-	if ((s = getenv("EXINIT")))
+	if ((s = getenv("EXINIT"))) {
 		ex_command(s)
+	} else {
+		char *homeenv = getenv("HOME");
+		if (homeenv) {
+			char exrc[PATH_MAX];
+			snprintf(exrc, sizeof(exrc), "%s/.exrc", homeenv);
+			load_exrc(exrc);
+		}
+	}
+	if (xexrc && strcmp(getcwd(NULL, 0), getenv("HOME")))
+		load_exrc(".exrc");
 }
diff --git a/vi.c b/vi.c
index 024601b..4c30f53 100644
--- a/vi.c
+++ b/vi.c
@@ -18,6 +18,7 @@
  */
 #include <ctype.h>
 #include <fcntl.h>
+#include <limits.h>
 #include <stdio.h>
 #include <stdarg.h>
 #include <stdlib.h>
