diff --git a/ex.c b/ex.c
index d45b663..fe974be 100644
--- a/ex.c
+++ b/ex.c
@@ -415,6 +415,23 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 			*ex_path ? ex_path : "unnamed", lbuf_len(xb),
 			fd < 0 || rd ? 'f' : 'r');
 	ex_show(msg);
+	if (!rd && fd >= 0 && lbuf_len(xb) > 0) {
+		int adv = 0;
+		while (lbuf_len(xb) > adv+1 && lbuf_buf(xb)[adv][0] == '\n')
+			adv++;
+		struct filetype lfts[] = {
+			{"sh", "^#!.*/(env[ \t]*)?(sh|bash|zsh|dash)([ \t]*.*)?$"},
+			{"py", "^#!.*/(env[ \t]*)?python3?([ \t]*.*)?$"}
+		};
+		char *pats[LEN(lfts)];
+		for (int i = 0; i < LEN(lfts); i++)
+			pats[i] = lfts[i].pat;
+		rset *rs = rset_make(LEN(lfts), pats, 0);
+		int hl = rset_find(rs, lbuf_buf(xb)[adv], 0, NULL, REG_NEWLINE);
+		if (hl >= 0)
+			ex_ft = syn_setft(lfts[hl].ft);
+		rset_free(rs);
+	}
 	return fd < 0 || rd;
 }
 
