diff --git a/ex.c b/ex.c
index e6613cd..6ad6dcb 100644
--- a/ex.c
+++ b/ex.c
@@ -418,6 +418,16 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 			*ex_path ? ex_path : "unnamed", lbuf_len(xb),
 			fd < 0 || rd ? 'f' : 'r');
 	ex_show(msg);
+	if (!rd && fd >= 0 && lbuf_len(xb) > 0) {
+		int adv = 0;
+		while (lbuf_len(xb) >= adv && lbuf_buf(xb)[adv][0] == '\n')
+			adv++;
+		char *langs[] = {"/bin/sh", "/bin/python"};
+		char *types[] = {"sh", "py"};
+		for (int i = 0; i < LEN(langs); i++)
+			if (strstr(lbuf_buf(xb)[adv], langs[i]))
+				ex_ft = syn_setft(types[i]);
+	}
 	return fd < 0 || rd;
 }
 
