diff --git a/ex.c b/ex.c
index 8f85bef..7741189 100644
--- a/ex.c
+++ b/ex.c
@@ -400,6 +400,16 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 		ex_show(msg);
 	} else
 		ex_show("read failed");
+	if (lbuf_len(xb) > 0) {
+		int adv = 0;
+		while (lbuf_len(xb) >= adv && lbuf_buf(xb)[adv][0] == '\n')
+			adv++;
+		char *langs[] = {"/bin/sh", "/bin/python"};
+		char *types[] = {"sh", "py"};
+		for (int i = 0; i < LEN(langs); i++)
+			if (strstr(lbuf_buf(xb)[adv], langs[i]))
+				ex_ft = syn_setft(types[i]);
+	}
 	return 0;
 }
 
