diff --git a/conf.c b/conf.c
index ddacc04e..641fd8a1 100644
--- a/conf.c
+++ b/conf.c
@@ -32,6 +32,7 @@ char ex_ft[] = "/ex";	/* ex mode (is never '\n' terminated) */
 char vs_ft[] = "/vs";	/* vi search prompt (is never '\n' terminated) */
 char bar_ft[] = "/-";	/* status bar (is never '\n' terminated) */
 char msg_ft[] = "/>";	/* ex message (is never '\n' terminated) */
+char grep_ft[] = "/g";	/* ex message (is never '\n' terminated) */
 
 #define IN	0	/* inverse | black */
 #define RE	1	/* red */
@@ -230,6 +231,11 @@ strike|tt|xmp|doctype|h1|h2|h3|h4|h5|h6|\
 		A(IN, SYN_BGMK(RE1), SYN_BGMK(AY1), SYN_BGMK(AY))},
 	{ac_ft, "[^ \t-/:-@[-^{-~]+$|(.+$)", A(IN, SYN_BGMK(AY1))},
 
+	/* grep buffer */
+	/* I don't feel like making new ->func index so reuse 3 here */
+	{grep_ft, "^(.+?):([0-9]+):(.+)", A(MA | SYN_SO, GR1, CY, AY1 | SYN_SO)},
+	{grep_ft, NULL, A(AY | SYN_BGMK(RE1)), 0, 3},
+
 	{ex_ft, "^[^:].*$", A(AY1 | SYN_BD)},
 	{ex_ft, "^.*$", A(AY1 | SYN_BD | SYN_SO)},
 	{ex_ft, ":[ \t]*((((?:[/?][^/?]*[/?]?)?[.%$]?(?:'[a-z'`[\\]*])?([0-9]*)?)\
diff --git a/ex.c b/ex.c
index 318aa4ec..dcb984be 100644
--- a/ex.c
+++ b/ex.c
@@ -33,7 +33,7 @@ sbuf *xacreg;			/* autocomplete db filter regex */
 rset *xkwdrs;			/* the last searched keyword rset */
 sbuf *xregs[256];		/* string registers */
 struct buf *bufs;		/* main buffers */
-struct buf tempbufs[2];		/* temporary buffers, for internal use */
+struct buf tempbufs[3];		/* temporary buffers, for internal use */
 struct buf *ex_buf;		/* current buffer */
 struct buf *ex_pbuf;		/* prev buffer */
 static struct buf *ex_tpbuf;	/* temp prev buffer */
diff --git a/vi.c b/vi.c
index 3b2a1fc4..94ffcb66 100644
--- a/vi.c
+++ b/vi.c
@@ -508,12 +508,12 @@ void dir_calc(char *path)
 	free(sb->s);
 }
 
-#define fssearch() \
+#define fssearch(isbuffer) \
 len = lbuf_s(path)->len; \
 path[len] = '\0'; \
 ret = ex_edit(path, len); \
 path[len] = '\n'; \
-if (ret && xrow) { \
+if (isbuffer) { \
 	*row = xrow; *off = xoff; /* short circuit */ \
 	if (!vi_search('n', cnt, row, off, 0)) \
 		return 1; \
@@ -524,14 +524,14 @@ if (ret && xrow) { \
 if (!vi_search(*row ? 'N' : 'n', cnt, row, off, 0)) \
 	return 1; \
 
-static int fs_search(int cnt, int *row, int *off)
+static int fs_search(int again, int cnt, int *row, int *off)
 {
 	char *path;
-	int again = 0, ret, len;
+	int ret, len;
 	wrap:
 	while (fspos < lbuf_len(tempbufs[1].lb)) {
 		path = tempbufs[1].lb->ln[fspos++];
-		fssearch()
+		fssearch(ret && xrow && again != 2)
 	}
 	if (fspos == lbuf_len(tempbufs[1].lb) && !again) {
 		fspos = 0;
@@ -547,7 +547,7 @@ static int fs_searchback(int cnt, int *row, int *off)
 	int ret, len;
 	while (--fspos >= 0) {
 		path = tempbufs[1].lb->ln[fspos];
-		fssearch()
+		fssearch(ret && xrow)
 	}
 	return 0;
 }
@@ -668,6 +668,7 @@ static int vi_motion(int *row, int *off)
 				break;
 		break;
 	case TK_CTL(']'):	/* note: this is also ^5 as per ascii */
+	case TK_CTL('x'):
 	case TK_CTL('p'):
 		#define open_saved(n) \
 		if (savepath[n]) { \
@@ -686,8 +687,44 @@ static int vi_motion(int *row, int *off)
 			lkwdcnt = xkwdcnt;
 			fspos += fsdir < 0 ? 1 : 0;
 			fspos = MIN(fspos, lbuf_len(tempbufs[1].lb));
-			fs_search(1, row, off);
+			fs_search(0, 1, row, off);
 			fsdir = 1;
+		} else if (mv == TK_CTL('x')) {
+			if (!strcmp(xb_path, "/grep/"))
+				temp_switch(2);
+			term_exec("", 1, '&')
+			temp_pos(2, -1, 0, 0);
+			sbuf *gstats; sbuf_make(gstats, 1024)
+			char nbuf[100];
+			int colpos = 0;
+			fspos = 0;
+			while (fs_search(2, 1, row, off)) {
+				sbuf_str(gstats, xb_path)
+				sbuf_chr(gstats, ':')
+				colpos = gstats->s_n;
+				#define nextrow() \
+				itoa((*row)+1, nbuf); \
+				sbuf_str(gstats, nbuf) \
+				sbuf_chr(gstats, ':') \
+				sbufn_str(gstats, lbuf_get(xb, *row)) \
+				temp_write(2, gstats->s); \
+				(*row)++; \
+
+				nextrow()
+				while (!vi_search('n', cnt, row, off, 0)) {
+					sbuf_cut(gstats, colpos)
+					nextrow()
+				}
+				sbuf_cut(gstats, 0)
+			}
+			sbuf_free(gstats)
+			for (i = xbufcur-1; i >= 0 && bufs[i].mtime == -1; i--)
+				ex_bufpostfix(&bufs[i], 1);
+			temp_switch(2);
+			lbuf_jump(xb, '[', row, off);
+			*off = 0;
+			syn_addhl(xregs['/'] ? xregs['/']->s : NULL, 3, 1); /* see conf.c */
+			syn_reloadft();
 		} else {
 			fspos -= fsdir > 0 ? 1 : 0;
 			if (!fs_searchback(1, row, off)) {
@@ -1264,7 +1301,17 @@ void vi(int init)
 				char buf[strlen(ln)+4];
 				strcpy(buf, ":e ");
 				strcpy(buf+3, ln);
-				term_push(buf, strlen(ln)+3);
+				if (!strcmp(xb_path, "/grep/")) {
+					int subs[2];
+					rset *rs = rset_make(1, (char*[]){":[0-9]+:"}, 0);
+					if (rset_find(rs, buf, subs, 0) >= 0) {
+						buf[subs[0]] = xsep;
+						buf[subs[1]-1] = '\n';
+						buf[subs[1]] = '\0';
+					}
+					rset_free(rs);
+				}
+				term_push(buf, strlen(buf));
 				break; }
 			case TK_CTL('n'):
 				vi_cndir = vi_arg ? -vi_cndir : vi_cndir;
@@ -1418,6 +1465,10 @@ void vi(int init)
 					}
 					ln = vi_enprompt(":", buf, &n);
 					goto do_excmd; }
+				case 'x':
+					temp_switch(2);
+					vi_mod = 1;
+					break;
 				default:
 					term_dec()
 				}
@@ -1788,6 +1839,7 @@ int main(int argc, char *argv[])
 	syn_init();
 	temp_open(0, "/hist/", _ft);
 	temp_open(1, "/fm/", fm_ft);
+	temp_open(2, "/grep/", grep_ft);
 	for (i = 1; i < argc && argv[i][0] == '-'; i++) {
 		if (argv[i][1] == '-' && !argv[i][2]) {
 			i++;
diff --git a/vi.h b/vi.h
index 2fe18c64..3f60664c 100644
--- a/vi.h
+++ b/vi.h
@@ -411,7 +411,7 @@ extern sbuf *xacreg;
 extern rset *xkwdrs;
 extern sbuf *xregs[256];
 extern struct buf *bufs;
-extern struct buf tempbufs[2];
+extern struct buf tempbufs[3];
 extern struct buf *ex_buf;
 extern struct buf *ex_pbuf;
 #define istempbuf(buf) (buf - bufs < 0 || buf - bufs >= xbufcur)
