diff --git a/conf.c b/conf.c
index 5e50d21..7fcb3b7 100644
--- a/conf.c
+++ b/conf.c
@@ -234,6 +234,11 @@ strike|tt|xmp|doctype|h1|h2|h3|h4|h5|h6|\
 		{0, SYN_BGMK(9), SYN_BGMK(8), SYN_BGMK(7)}},
 	{"/ac", "[^ \t-/:-@[-^{-~]+$|(.+$)", {0, SYN_BGMK(8)}},
 
+	/* grep buffer */
+	/* I don't feel like making new ->func index so reuse 3 here */
+	{"/grep", "^(.+?):([0-9]+):(.+)", {5, 10, 6, 8}, {1, 0, 0, 1}},
+	{"/grep", NULL, {7 | SYN_BGMK(9)}, {0}, 0, 3},
+
 	/* status bar (is never '\n' terminated) */
 	{"/-", "^(\".*\").*(\\[[wrf]\\]).*$", {8 | SYN_BD, 4, 1}},
 	{"/-", "^(\".*\").* ([0-9]{1,3}%) (L[0-9]+) (C[0-9]+) (B-?[0-9]+)?.*$",
diff --git a/ex.c b/ex.c
index 8acbc66..b10d460 100644
--- a/ex.c
+++ b/ex.c
@@ -22,7 +22,7 @@ int xpac;			/* print autocomplete options */
 int xkwdcnt;			/* number of search kwd changes */
 int xbufcur;			/* number of active buffers */
 struct buf *bufs;		/* main buffers */
-struct buf tempbufs[2];		/* temporary buffers, for internal use */
+struct buf tempbufs[3];		/* temporary buffers, for internal use */
 struct buf *ex_buf;		/* current buffer */
 struct buf *ex_pbuf;		/* prev buffer */
 static struct buf *ex_tpbuf;	/* temp prev buffer */
diff --git a/vi.c b/vi.c
index 869b920..a9a9099 100644
--- a/vi.c
+++ b/vi.c
@@ -657,12 +657,12 @@ void dir_calc(char *path)
 	}
 }
 
-#define fssearch() \
+#define fssearch(isbuffer) \
 len = lbuf_slen(path); \
 path[len] = '\0'; \
 ret = ex_edit(path, len); \
 path[len] = '\n'; \
-if (ret && xrow) { \
+if (isbuffer) { \
 	*row = xrow; *off = xoff; /* short circuit */ \
 	if (!vi_search('n', cnt, row, off, 0)) \
 		return 1; \
@@ -673,14 +673,14 @@ if (ret && xrow) { \
 if (!vi_search(*row ? 'N' : 'n', cnt, row, off, 0)) \
 	return 1; \
 
-static int fs_search(int cnt, int *row, int *off)
+static int fs_search(int again, int cnt, int *row, int *off)
 {
 	char *path;
-	int again = 0, ret, len;
+	int ret, len;
 	wrap:
 	while (fspos < lbuf_len(tempbufs[1].lb)) {
 		path = lbuf_get(tempbufs[1].lb, fspos++);
-		fssearch()
+		fssearch(ret && xrow && again != 2)
 	}
 	if (fspos == lbuf_len(tempbufs[1].lb) && !again) {
 		fspos = 0;
@@ -696,7 +696,7 @@ static int fs_searchback(int cnt, int *row, int *off)
 	int ret, len;
 	while (--fspos >= 0) {
 		path = lbuf_get(tempbufs[1].lb, fspos);
-		fssearch()
+		fssearch(ret && xrow)
 	}
 	return 0;
 }
@@ -799,6 +799,7 @@ static int vi_motion(int *row, int *off)
 				break;
 		break;
 	case TK_CTL(']'):	/* note: this is also ^5 as per ascii */
+	case TK_CTL('x'):
 	case TK_CTL('p'):
 		#define open_saved(n) \
 		if (savepath[n]) { \
@@ -817,8 +818,44 @@ static int vi_motion(int *row, int *off)
 			lkwdcnt = xkwdcnt;
 			fspos += fsdir < 0 ? 1 : 0;
 			fspos = MIN(fspos, lbuf_len(tempbufs[1].lb));
-			fs_search(1, row, off);
+			fs_search(0, 1, row, off);
 			fsdir = 1;
+		} else if (mv == TK_CTL('x')) {
+			if (!strcmp(ex_path, "/grep/"))
+				temp_switch(2);
+			term_exec("qq", 3, /*nop*/, /*nop*/)
+			temp_pos(2, -1, 0, 0);
+			sbuf *gstats; sbuf_make(gstats, 1024)
+			char nbuf[100];
+			int colpos = 0;
+			fspos = 0;
+			while (fs_search(2, 1, row, off)) {
+				sbuf_str(gstats, ex_path)
+				sbuf_chr(gstats, ':')
+				colpos = gstats->s_n;
+				#define nextrow() \
+				itoa((*row)+1, nbuf); \
+				sbuf_str(gstats, nbuf) \
+				sbuf_chr(gstats, ':') \
+				sbufn_str(gstats, lbuf_get(xb, *row)) \
+				temp_write(2, gstats->s); \
+				(*row)++; \
+
+				nextrow()
+				while (!vi_search('n', cnt, row, off, 0)) {
+					sbuf_cut(gstats, colpos)
+					nextrow()
+				}
+				sbuf_cut(gstats, 0)
+			}
+			sbuf_free(gstats)
+			for (i = xbufcur-1; i >= 0 && bufs[i].mtime == -1; i--)
+				ex_bufpostfix(&bufs[i], 1);
+			temp_switch(2);
+			lbuf_jump(xb, '[', row, off);
+			*off = 0;
+			syn_addhl(xregs['/'], 3, 1); /* see conf.c */
+			syn_reloadft();
 		} else {
 			fspos -= fsdir > 0 ? 1 : 0;
 			if (!fs_searchback(1, row, off)) {
@@ -1500,7 +1537,17 @@ void vi(int init)
 				buf[1] = 'e';
 				buf[2] = ' ';
 				strcpy(buf+3, ln);
-				term_push(buf, strlen(ln)+3);
+				if (!strcmp(ex_path, "/grep/")) {
+					int subs[2];
+					rset *rs = rset_make(1, (char*[]){":[0-9]+:"}, 0);
+					if (rset_find(rs, buf, 1, subs, 0) >= 0) {
+						buf[subs[0]] = '|';
+						buf[subs[1]-1] = '\n';
+						buf[subs[1]] = '\0';
+					}
+					rset_free(rs);
+				}
+				term_push(buf, strlen(buf));
 				break; }
 			case TK_CTL('n'):
 				vi_cndir = vi_arg1 ? -vi_cndir : vi_cndir;
@@ -1653,6 +1700,10 @@ void vi(int init)
 					}
 					ln = vi_prompt(":", buf, &kmap);
 					goto do_excmd; }
+				case 'x':
+					temp_switch(2);
+					vi_mod = 1;
+					break;
 				default:
 					vi_back(k);
 				}
@@ -2020,6 +2071,7 @@ int main(int argc, char *argv[])
 	syn_init();
 	temp_open(0, "/hist/", "/");
 	temp_open(1, "/fm/", "/fm");
+	temp_open(2, "/grep/", "/grep");
 	for (i = 1; i < argc && argv[i][0] == '-'; i++) {
 		if (argv[i][1] == '-' && !argv[i][2]) {
 			i++;
diff --git a/vi.h b/vi.h
index 1d23d99..5e76d15 100644
--- a/vi.h
+++ b/vi.h
@@ -316,7 +316,7 @@ extern int xbufcur;
 extern struct buf *ex_buf;
 extern struct buf *ex_pbuf;
 extern struct buf *bufs;
-extern struct buf tempbufs[2];
+extern struct buf tempbufs[3];
 #define istempbuf(buf) (buf - bufs < 0 || buf - bufs >= xbufcur)
 #define EXLEN	512	/* ex line length */
 #define ex_path ex_buf->path
