diff --git a/ex.c b/ex.c
index 2c841d1..8a6c1b3 100644
--- a/ex.c
+++ b/ex.c
@@ -9,6 +9,7 @@ int xhll;			/* highlight current line */
 int xhlw;			/* highlight current word */
 int xhlp;			/* highlight {}[]() pair */
 int xhlr;			/* highlight text in reverse direction */
+int xhlm;			/* highlight marks */
 int xled = 1;			/* use the line editor */
 int xtd = +1;			/* current text direction */
 int xshape = 1;			/* perform letter shaping */
@@ -851,6 +852,7 @@ static struct option {
 	{"hlw", &xhlw},
 	{"hlp", &xhlp},
 	{"hlr", &xhlr},
+	{"hlm", &xhlm},
 	{"tbs", &xtabspc},
 	{"ish", &xish},
 	{"grp", &xgrp},
diff --git a/led.c b/led.c
index 864cd94..38071bd 100644
--- a/led.c
+++ b/led.c
@@ -130,6 +130,21 @@ static void led_markrev(int n, char **chrs, int *pos, int *att)
 	}
 }
 
+static void led_hlmark(int *att, int row)
+{
+	int mrow = -1, moff, icol = 0;
+	char marks[] = "abcdefghijklmnopqrstuvwxyz[]`'";
+	for (int i = 0; i < LEN(marks); icol++, i++) {
+		if (lbuf_jump(xb, marks[i], &mrow, &moff) || mrow != row)
+			continue;
+		if (icol > 15) /* wrap for only 16 color term */
+			icol = 0;
+		att[moff] = SYN_BGMK(icol+1);
+		mrow = -1;
+	}
+	
+}
+
 static char *led_bounds(int *off, char **chrs, int cterm)
 {
 	int l, i = 0;
@@ -257,6 +272,8 @@ void led_render(char *s0, int row, int cbeg, int cend)
 	}
 	if (xhlr)
 		led_markrev(n, chrs, pos, ratt);
+	if (xhlm && row >= 0)
+		led_hlmark(ratt, row+xtop);
 	/* generate term output */
 	term_pos(row, 0);
 	term_kill();
