diff --git a/ex.c b/ex.c
index e7c1a02..b7d9b58 100644
--- a/ex.c
+++ b/ex.c
@@ -9,6 +9,7 @@ int xhll;			/* highlight current line */
 int xhlw;			/* highlight current word */
 int xhlp;			/* highlight {}[]() pair */
 int xhlr;			/* highlight text in reverse direction */
+int xhlm;			/* highlight marks */
 int xled = 1;			/* use the line editor */
 int xtd = +1;			/* current text direction */
 int xshape = 1;			/* perform letter shaping */
@@ -889,6 +890,7 @@ static struct option {
 	{"hlw", &xhlw},
 	{"hlp", &xhlp},
 	{"hlr", &xhlr},
+	{"hlm", &xhlm},
 	{"tbs", &xtabspc},
 	{"ish", &xish},
 	{"grp", &xgrp},
diff --git a/vi.c b/vi.c
index 7d5c53a..ee104c7 100644
--- a/vi.c
+++ b/vi.c
@@ -1834,6 +1834,21 @@ void vi(int init)
 				}
 			}
 		}
+		if (xhlm) {
+			int mrow, moff;
+			char marks[] = "abcdefghijklmnopqrstuvwxyz[]`'";
+			led_att la;
+			if (!led_attsb)
+				sbuf_make(led_attsb, sizeof(la))
+			for (int i = 0; i < LEN(marks); i++) {
+				if (lbuf_jump(xb, marks[i], &mrow, &moff))
+					continue;
+				la.s = lbuf_get(xb, mrow);
+				la.off = moff;
+				la.att = SYN_BGMK((i % 15) + 1);
+				sbuf_mem(led_attsb, &la, (int)sizeof(la))
+			}
+		}
 		if (xhlp && (k = syn_findhl(3)) >= 0) {
 			int row = xrow, off = xoff;
 			int row1 = xrow, off1 = xoff;
