diff --git a/ex.c b/ex.c
index 5bf89e3b..8c74a1ee 100644
--- a/ex.c
+++ b/ex.c
@@ -9,6 +9,7 @@ int xhll;			/* highlight current line */
 int xhlw;			/* highlight current word */
 int xhlp;			/* highlight {}[]() pair */
 int xhlr;			/* highlight text in reverse direction */
+int xhlm;			/* highlight marks */
 int xled = 1;			/* use the line editor */
 int xtd = +1;			/* current text direction */
 int xshape = 1;			/* perform letter shaping */
@@ -1098,6 +1099,7 @@ static int eo_##opt(char *loc, char *cmd, char *arg) { inner }
 EO(pac) EO(pr) EO(ai) EO(ish) EO(ic) EO(grp) EO(shape) EO(seq)
 EO(sep) EO(tbs) EO(td) EO(order) EO(hll) EO(hlw) EO(hlp) EO(hlr)
 EO(hl) EO(lim) EO(led) EO(vis) EO(mpt)
+EO(hlm)
 
 _EO(left,
 	if (*loc)
@@ -1175,6 +1177,7 @@ static struct excmd {
 	EO(tbs),
 	EO(td),
 	EO(order),
+	EO(hlm),
 	EO(hll),
 	EO(hlw),
 	EO(hlp),
diff --git a/vi.c b/vi.c
index 98367572..21e299ee 100644
--- a/vi.c
+++ b/vi.c
@@ -1795,6 +1795,21 @@ void vi(int init)
 				}
 			}
 		}
+		if (xhlm) {
+			int mrow, moff;
+			char marks[] = "abcdefghijklmnopqrstuvwxyz[]`'";
+			led_att la;
+			if (!led_attsb)
+				sbuf_make(led_attsb, sizeof(la))
+			for (int i = 0; i < LEN(marks); i++) {
+				if (lbuf_jump(xb, marks[i], &mrow, &moff))
+					continue;
+				la.s = lbuf_get(xb, mrow);
+				la.off = moff;
+				la.att = SYN_BGMK(((i % 15) + 1));
+				sbuf_mem(led_attsb, &la, (int)sizeof(la))
+			}
+		}
 		if (xhlp && (k = syn_findhl(3)) >= 0) {
 			int row = xrow, off = xoff;
 			int row1 = xrow, off1 = xoff;
