diff --git a/kmap.h b/kmap.h
index d025f5f..e549d13 100644
--- a/kmap.h
+++ b/kmap.h
@@ -1,5 +1,13 @@
 static char *kmap_en[256] = {
 	[0] = "en",
+	['y'] = "h",
+	['n'] = "j",
+	['e'] = "k",
+	['o'] = "l",
+	['h'] = "y",
+	['j'] = "n",
+	['k'] = "e",
+	['l'] = "o",
 };
 
 static char *kmap_fa[256] = {
diff --git a/term.c b/term.c
index 7a761c1..4ee0ab4 100644
--- a/term.c
+++ b/term.c
@@ -151,10 +151,14 @@ int term_read(void)
 			err:
 			*ibuf = 0;
 		} else if (texec && ibuf_pos < sizeof(ibuf)) {
+			if (*ibuf > 0 && conf_kmap(0)[*ibuf])
+				*ibuf = *conf_kmap(0)[*ibuf];
 			ibuf_cnt++;
 			ibuf[ibuf_pos] = *ibuf;
 			goto ret;
 		}
+		if (*ibuf > 0 && conf_kmap(0)[*ibuf])
+			*ibuf = *conf_kmap(0)[*ibuf];
 		ibuf_cnt = 1;
 		ibuf_pos = 0;
 	}
