diff --git a/kmap.h b/kmap.h
index d9f602e..c0a6533 100644
--- a/kmap.h
+++ b/kmap.h
@@ -1,5 +1,13 @@
 static char *kmap_en[256] = {
 	[0] = "en",
+	['y'] = "h",
+	['n'] = "j",
+	['e'] = "k",
+	['o'] = "l",
+	['h'] = "y",
+	['j'] = "n",
+	['k'] = "e",
+	['l'] = "o",
 };
 
 static char *kmap_fa[256] = {
diff --git a/term.c b/term.c
index 4d2e41a..f4adfda 100644
--- a/term.c
+++ b/term.c
@@ -134,6 +134,10 @@ int term_read(void)
 		/* read a single input character */
 		if ((n = read(STDIN_FILENO, ibuf, 1)) <= 0)
 			return -1;
+		if (conf_kmap(0)[(unsigned char)*ibuf]) {
+			n = MIN(strlen(conf_kmap(0)[(unsigned char)*ibuf]), sizeof(ibuf)-1);
+			memcpy(ibuf, conf_kmap(0)[(unsigned char)*ibuf], n);
+		}
 		ibuf_cnt = n;
 		ibuf_pos = 0;
 	}
