diff --git a/ex.c b/ex.c
index c5b722d..5139cd3 100644
--- a/ex.c
+++ b/ex.c
@@ -181,6 +181,9 @@ static char *ex_pathexpand(char *src, int spaceallowed)
 			}
 			sbuf_str(sb, pbuf->path)
 			src += n >= 0 ? snprintf(0, 0, "%+d", n) : 1;
+		} else if (*src == '~') {
+			sbuf_str(sb, getenv("HOME"))
+			src++;
 		} else {
 			if (*src == '\\' && src[1])
 				src++;
diff --git a/vi.c b/vi.c
index d4960c4..7bf0209 100644
--- a/vi.c
+++ b/vi.c
@@ -1679,6 +1679,12 @@ void vi(int init)
 					free(ln);
 					free(cs);
 					break;
+				case 'n':
+					k = vi_read();
+					char buf[] = "i ";
+					buf[1] = k;
+					term_push(buf, 3);
+					break;
 				case 't': {
 					strcpy(vi_msg, "arg2:(0|#)");
 					vi_drawmsg();
