diff --git a/ex.c b/ex.c
index d29c595..a45649e 100644
--- a/ex.c
+++ b/ex.c
@@ -16,6 +16,7 @@ int xorder = 1;			/* change the order of characters */
 int xkmap;			/* the current keymap */
 int xkmap_alt = 1;		/* the alternate keymap */
 int xtabspc = 8;		/* number of spaces for tab */
+int xqexit = 1000;		/* exit insert via kj (delay in ms) */
 int xish = 1;			/* interactive shell */
 int xgrp;			/* regex search group */
 int xpac;			/* print autocomplete options */
@@ -860,6 +861,7 @@ static struct option {
 	{"hlp", &xhlp},
 	{"hlr", &xhlr},
 	{"tbs", &xtabspc},
+	{"qe", &xqexit},
 	{"ish", &xish},
 	{"grp", &xgrp},
 	{"pac", &xpac},
diff --git a/led.c b/led.c
index abf62fd..c3ee3d1 100644
--- a/led.c
+++ b/led.c
@@ -376,6 +376,14 @@ static void led_redraw(char *cs, int r, int orow, int lsh)
 	term_pos(xrow - xtop, 0);
 }
 
+static int gettime_ms(void)
+{
+	struct timespec t;
+	if (clock_gettime(CLOCK_MONOTONIC, &t) < 0)
+		return 0;
+	return t.tv_sec * 1000 + t.tv_nsec / 1000000;
+}
+
 /* read a line from the terminal */
 static void led_line(sbuf *sb, int ps, int pre, char *post, int ai_max,
 		int *key, int *kmap, int orow, int lsh)
@@ -383,6 +391,7 @@ static void led_line(sbuf *sb, int ps, int pre, char *post, int ai_max,
 	int len, p_reg = 0;
 	int c, i = 0, last_sug = 0, sug_pt = -1;
 	char *cs, *sug = NULL, *_sug = NULL;
+	int quickexit = 0;
 	while (1) {
 		led_printparts(sb, ps, post, ai_max);
 		len = sb->s_n;
@@ -580,7 +589,21 @@ static void led_line(sbuf *sb, int ps, int pre, char *post, int ai_max,
 			xquit = 0;
 			restore(xvis)
 			continue;
+		case 'j':
+			if (xqexit && (gettime_ms() - quickexit) < xqexit) {
+				if (sb->s_n) {
+					if (sb->s[led_lastchar(sb->s)] != 'k')
+						goto _default;
+					sbuf_cut(sb, led_lastchar(sb->s))
+				}
+				c = TK_ESC;
+				goto leave;
+			}
+			goto _default;
+		case 'k':
+			quickexit = gettime_ms();
 		default:
+_default:
 			if (c == '\n' || TK_INT(c))
 				goto leave;
 			if ((cs = led_read(kmap, c)))
diff --git a/vi.c b/vi.c
index 83c51ad..976f18b 100644
--- a/vi.c
+++ b/vi.c
@@ -25,6 +25,7 @@
 #include <dirent.h>
 #include <signal.h>
 #include <unistd.h>
+#include <time.h>
 #include <poll.h>
 #include <termios.h>
 #include <sys/stat.h>
diff --git a/vi.h b/vi.h
index 3a96ecb..3fecf61 100644
--- a/vi.h
+++ b/vi.h
@@ -440,6 +440,7 @@ extern int xhlr;
 extern int xkmap;
 extern int xkmap_alt;
 extern int xtabspc;
+extern int xqexit;
 extern int xish;
 extern int xgrp;
 extern int xpac;
