diff --git a/ex.c b/ex.c
index e249b726..62277baa 100644
--- a/ex.c
+++ b/ex.c
@@ -13,6 +13,7 @@ int xtd = +1;			/* current text direction */
 int xshape = 1;			/* perform letter shaping */
 int xorder = 1;			/* change the order of characters */
 int xtbs = 8;			/* number of spaces for tab */
+int xqe = 1000;			/* exit insert via kj (delay in ms) */
 int xish;			/* interactive shell */
 int xgrp;			/* regex search group */
 int xpac;			/* print autocomplete options */
@@ -1292,6 +1293,7 @@ static void *eo_##opt(char *loc, char *cmd, char *arg) { inner }
 EO(pac) EO(pr) EO(ai) EO(ish) EO(ic) EO(grp) EO(shape) EO(seq)
 EO(sep) EO(tbs) EO(td) EO(order) EO(hll) EO(hlw) EO(hlp) EO(hlr)
 EO(hl) EO(lim) EO(led) EO(vis) EO(mpt) EO(err)
+EO(qe)
 
 _EO(left,
 	if (*loc)
@@ -1345,6 +1347,7 @@ static struct excmd {
 	{"g", ec_glob},
 	EO(mpt),
 	{"m", ec_mark},
+	EO(qe),
 	{"q!", ec_quit},
 	{"q", ec_quit},
 	{"reg", ec_regprint},
diff --git a/led.c b/led.c
index 42ebdc54..06c03c5b 100644
--- a/led.c
+++ b/led.c
@@ -404,6 +404,14 @@ void led_modeswap(void)
 	restore(xvis)
 }
 
+static int gettime_ms(void)
+{
+	struct timespec t;
+	if (clock_gettime(CLOCK_MONOTONIC, &t) < 0)
+		return 0;
+	return t.tv_sec * 1000 + t.tv_nsec / 1000000;
+}
+
 /* read a line from the terminal */
 static void led_line(sbuf *sb, int ps, int pre, char **post, int postn, char **postref,
 	int ai_max, int *key, int *kmap, int orow, int crow, int ctop, int flg)
@@ -411,6 +419,7 @@ static void led_line(sbuf *sb, int ps, int pre, char **post, int postn, char **p
 	int len, t_row = -2, p_reg = 0;
 	int c, i, lsug = 0, sug_pt = -1;
 	char *cs, *sug = NULL, *_sug = NULL;
+	int quickexit = 0;
 	do {
 		led_printparts(sb, pre, ps, *post, postn, ai_max);
 		len = sb->s_n;
@@ -627,7 +636,21 @@ static void led_line(sbuf *sb, int ps, int pre, char **post, int postn, char **p
 				exbuf_load(ex_buf)
 			}
 			continue; }
+		case 'j':
+			if (xqe && (gettime_ms() - quickexit) < xqe) {
+				if (sb->s_n) {
+					if (sb->s[led_lastchar(sb->s)] != 'k')
+						goto _default;
+					sbuf_cut(sb, led_lastchar(sb->s))
+				}
+				c = TK_ESC;
+				goto leave;
+			}
+			goto _default;
+		case 'k':
+			quickexit = gettime_ms();
 		default:
+_default:
 			if (c == '\n' || TK_INT(c))
 				goto leave;
 			if ((cs = led_read(kmap, c)))
diff --git a/vi.c b/vi.c
index d7068dfc..351e04dc 100644
--- a/vi.c
+++ b/vi.c
@@ -7,6 +7,7 @@
 #include <dirent.h>
 #include <signal.h>
 #include <unistd.h>
+#include <time.h>
 #include <poll.h>
 #include <termios.h>
 #include <limits.h>
@@ -1505,6 +1506,8 @@ void vi(int init)
 				vc_insert(c);
 				ins:
 				vi_mod |= !xpac && xrow == orow ? 8 : 1;
+				if (xqe)
+					vi_mod |= 2;
 				if (vi_insmov == 127) {
 					if (xrow && !(xoff > 0 && lbuf_eol(xb, xrow, 1))) {
 						xrow--;
diff --git a/vi.h b/vi.h
index 23162165..118eec48 100644
--- a/vi.h
+++ b/vi.h
@@ -406,6 +406,7 @@ extern int xtd;
 extern int xshape;
 extern int xorder;
 extern int xtbs;
+extern int xqe;
 extern int xish;
 extern int xgrp;
 extern int xpac;
