diff --git a/ex.c b/ex.c
index d45b663..706acff 100644
--- a/ex.c
+++ b/ex.c
@@ -34,6 +34,7 @@ static int xbufsmax;		/* number of buffers */
 static int xbufsalloc = 10;	/* initial number of buffers */
 static char xrep[EXLEN];	/* the last replacement */
 static int xgdep;		/* global command recursion depth */
+char readonly = 0;		/* commandline readonly option */
 
 static int rstrcmp(const char *s1, const char *s2, int l1, int l2)
 {
@@ -95,6 +96,7 @@ static int bufs_open(const char *path, int len)
 	bufs[i].top = 0;
 	bufs[i].td = +1;
 	bufs[i].mtime = -1;
+	bufs[i].readonly = readonly;
 	return i;
 }
 
@@ -406,6 +408,8 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 		bufs_switch(bufs_open(arg+cd, len));
 		cd = 3; /* XXX: sigh... */
 	}
+	if (access(arg, F_OK) == 0 && access(arg, W_OK) == -1)
+		ex_buf->readonly = 1;
 	readfile(rd =)
 	if (cd == 3 || (!rd && fd >= 0)) {
 		ex_bufpostfix(ex_buf, arg[0]);
@@ -516,14 +520,19 @@ static int ec_write(char *loc, char *cmd, char *arg)
 		free(ibuf);
 	} else {
 		int fd;
-		if (!strchr(cmd, '!') && !strcmp(ex_path, path) &&
-				mtime(ex_path) > ex_buf->mtime) {
-			ex_show("write failed: file changed");
-			return 1;
-		}
-		if (!strchr(cmd, '!') && arg[0] && mtime(arg) >= 0) {
-			ex_show("write failed: file exists");
-			return 1;
+		if (!strchr(cmd, '!')) {
+			if (ex_buf->readonly) {
+				ex_show("write failed: readonly option is set");
+				return 1;
+			}
+			if (!strcmp(ex_path, path) && mtime(ex_path) > ex_buf->mtime) {
+				ex_show("write failed: file changed");
+				return 1;
+			}
+			if (arg[0] && mtime(arg) >= 0) {
+				ex_show("write failed: file exists");
+				return 1;
+			}
 		}
 		fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, conf_mode());
 		if (fd < 0) {
@@ -989,6 +998,12 @@ static int ec_setbufsmax(char *loc, char *cmd, char *arg)
 	return 0;
 }
 
+static int ec_readonly(char *loc, char *cmd, char *arg)
+{
+	ex_buf->readonly = !ex_buf->readonly;
+	return 0;
+}
+
 static struct excmd {
 	char *name;
 	int (*ec)(char *loc, char *cmd, char *arg);
@@ -1037,6 +1052,7 @@ static struct excmd {
 	{"inc", ec_setincl},
 	{"bx", ec_setbufsmax},
 	{"ac", ec_setacreg},
+	{"ro", ec_readonly},
 	{"", ec_print},
 };
 
diff --git a/vi.c b/vi.c
index c4d046a..028b6a5 100644
--- a/vi.c
+++ b/vi.c
@@ -2056,9 +2056,11 @@ int main(int argc, char *argv[])
 				xvis |= 4;
 			else if (argv[i][j] == 'v')
 				xvis &= ~4;
+			else if (argv[i][j] == 'R')
+				readonly = 1;
 			else {
 				fprintf(stderr, "Unknown option: -%c\n", argv[i][j]);
-				fprintf(stderr, "Usage: %s [-esv] [file ...]\n", argv[0]);
+				fprintf(stderr, "Usage: %s [-eRsv] [file ...]\n", argv[0]);
 				return EXIT_FAILURE;
 			}
 		}
diff --git a/vi.h b/vi.h
index 00ba570..eacd181 100644
--- a/vi.h
+++ b/vi.h
@@ -297,6 +297,7 @@ struct buf {
 	int plen, row, off, top;
 	long mtime;			/* modification time */
 	signed char td;			/* text direction */
+	char readonly;			/* read only */
 };
 extern int xbufcur;
 extern struct buf *ex_buf;
@@ -435,3 +436,4 @@ extern rset *fsincl;
 extern char *fs_exdir;
 extern int vi_hidch;
 extern int vi_insmov;
+extern char readonly;
