diff --git a/ex.c b/ex.c
index 168c0b01..10e33e3c 100644
--- a/ex.c
+++ b/ex.c
@@ -45,6 +45,7 @@ static char xserr[] = "syntax error";
 static char xirerr[] = "invalid range";
 static char xsrerr[] = "range not found";
 static char *xrerr;
+char readonly = 0;		/* commandline readonly option */
 
 static int rstrcmp(const char *s1, const char *s2, int l1, int l2)
 {
@@ -106,6 +107,7 @@ static int bufs_open(const char *path, int len)
 	bufs[i].top = 0;
 	bufs[i].td = +1;
 	bufs[i].mtime = -1;
+	bufs[i].readonly = readonly;
 	return i;
 }
 
@@ -432,6 +434,8 @@ static void *ec_edit(char *loc, char *cmd, char *arg)
 		bufs_switch(bufs_open(arg+cd, len));
 		cd = 3; /* XXX: quick hack to indicate new lbuf */
 	}
+	if (access(arg, F_OK) == 0 && access(arg, W_OK) == -1)
+		ex_buf->readonly = 1;
 	readfile(rd =)
 	if (cd == 3 || (!rd && fd >= 0)) {
 		ex_bufpostfix(ex_buf, arg[0]);
@@ -593,6 +597,8 @@ static void *ec_write(char *loc, char *cmd, char *arg)
 		free(ibuf.s);
 	} else {
 		if (!strchr(cmd, '!')) {
+			if (ex_buf->readonly)
+				return "write failed: readonly option is set";
 			if (!strcmp(xb_path, path) && mtime(path) > ex_buf->mtime)
 				return "write failed: file changed";
 			if (arg[0] && mtime(path) >= 0)
@@ -1182,6 +1188,12 @@ static void *ec_setenc(char *loc, char *cmd, char *arg)
 	return NULL;
 }
 
+static void *ec_readonly(char *loc, char *cmd, char *arg)
+{
+	ex_buf->readonly = !ex_buf->readonly;
+	return NULL;
+}
+
 static int eo_val(char *arg)
 {
 	int val = atoi(arg);
@@ -1254,6 +1266,7 @@ static struct excmd {
 	{"q", ec_quit},
 	{"reg", ec_regprint},
 	{"rd", ec_undoredo},
+	{"ro", ec_readonly},
 	{"r", ec_read},
 	{"wq!", ec_write},
 	{"wq", ec_write},
diff --git a/vi.c b/vi.c
index cbad0b34..6c8a96b3 100644
--- a/vi.c
+++ b/vi.c
@@ -1838,9 +1838,11 @@ int main(int argc, char *argv[])
 				xvis |= 8;
 			else if (argv[i][j] == 'v')
 				xvis &= ~4;
+			else if (argv[i][j] == 'R')
+				readonly = 1;
 			else {
 				fprintf(stderr, "Unknown option: -%c\n", argv[i][j]);
-				fprintf(stderr, "Nextvi-2.5 Usage: %s [-emsv] [file ...]\n", argv[0]);
+				fprintf(stderr, "Nextvi-2.5 Usage: %s [-emRsv] [file ...]\n", argv[0]);
 				return EXIT_FAILURE;
 			}
 		}
diff --git a/vi.h b/vi.h
index c94de90d..0e9a84c2 100644
--- a/vi.h
+++ b/vi.h
@@ -387,6 +387,7 @@ struct buf {
 	int plen, row, off, top;
 	long mtime;			/* modification time */
 	signed char td;			/* text direction */
+	char readonly;			/* read only */
 };
 /* ex options */
 extern int xleft;
@@ -520,3 +521,4 @@ extern char vi_msg[512];
 extern rset *fsincl;
 extern char *fs_exdir;
 void dir_calc(char *path);
+extern char readonly;
