diff --git a/ex.c b/ex.c
index 19ede9aa..55ef44f2 100644
--- a/ex.c
+++ b/ex.c
@@ -45,6 +45,7 @@ static char xserr[] = "syntax error";
 static char xirerr[] = "invalid range";
 static char xsrerr[] = "range not found";
 static char *xrerr;
+char readonly = 0;		/* commandline readonly option */
 
 static int rstrcmp(const char *s1, const char *s2, int l1, int l2)
 {
@@ -106,6 +107,7 @@ static int bufs_open(const char *path, int len)
 	bufs[i].top = 0;
 	bufs[i].td = +1;
 	bufs[i].mtime = -1;
+	bufs[i].readonly = readonly;
 	return i;
 }
 
@@ -439,6 +441,8 @@ static void *ec_edit(char *loc, char *cmd, char *arg)
 		ex_bufpostfix(ex_buf, arg[0]);
 		syn_setft(xb_ft);
 	}
+	if (access(arg, F_OK) == 0 && access(arg, W_OK) == -1)
+		ex_buf->readonly = 1;
 	snprintf(msg, sizeof(msg), "\"%s\" %dL [%c]",
 			*xb_path ? xb_path : "unnamed", lbuf_len(xb),
 			fd < 0 || rd ? 'f' : 'r');
@@ -572,6 +576,8 @@ static void *ec_write(char *loc, char *cmd, char *arg)
 	} else {
 		int fd;
 		if (!strchr(cmd, '!')) {
+			if (ex_buf->readonly)
+				return "write failed: readonly option is set";
 			if (!strcmp(xb_path, path) && mtime(path) > ex_buf->mtime)
 				return "write failed: file changed";
 			if (arg[0] && mtime(path) >= 0)
@@ -1175,6 +1181,12 @@ static void *ec_setenc(char *loc, char *cmd, char *arg)
 	return NULL;
 }
 
+static void *ec_readonly(char *loc, char *cmd, char *arg)
+{
+	ex_buf->readonly = !ex_buf->readonly;
+	return NULL;
+}
+
 static int eo_val(char *arg)
 {
 	int val = atoi(arg);
@@ -1246,6 +1258,7 @@ static struct excmd {
 	{"q", ec_quit},
 	{"reg", ec_regprint},
 	{"rd", ec_undoredo},
+	{"ro", ec_readonly},
 	{"r", ec_read},
 	{"wq!", ec_write},
 	{"wq", ec_write},
diff --git a/vi.c b/vi.c
index e1f1cb55..99922b16 100644
--- a/vi.c
+++ b/vi.c
@@ -1799,9 +1799,11 @@ int main(int argc, char *argv[])
 				xvis |= 8;
 			else if (argv[i][j] == 'v')
 				xvis &= ~4;
+			else if (argv[i][j] == 'R')
+				readonly = 1;
 			else {
 				fprintf(stderr, "Unknown option: -%c\n", argv[i][j]);
-				fprintf(stderr, "Usage: %s [-emsv] [file ...]\n", argv[0]);
+				fprintf(stderr, "Usage: %s [-emRsv] [file ...]\n", argv[0]);
 				return EXIT_FAILURE;
 			}
 		}
diff --git a/vi.h b/vi.h
index 964e4b1d..3458f65f 100644
--- a/vi.h
+++ b/vi.h
@@ -374,6 +374,7 @@ struct buf {
 	int plen, row, off, top;
 	long mtime;			/* modification time */
 	signed char td;			/* text direction */
+	char readonly;			/* read only */
 };
 /* ex options */
 extern int xleft;
@@ -508,3 +509,4 @@ extern char vi_msg[512];
 extern rset *fsincl;
 extern char *fs_exdir;
 void dir_calc(char *path);
+extern char readonly;
