diff --git a/ex.c b/ex.c
index 86fd431b..165c7e95 100644
--- a/ex.c
+++ b/ex.c
@@ -22,6 +22,7 @@ int xlim = -1;			/* rendering cutoff for non cursor lines */
 int xseq = 1;			/* undo/redo sequence */
 int xerr = 1;			/* error handling -
 				bit 1: print errors, bit 2: early return, bit 3: ignore errors */
+int xrec;			/* input recoding register */
 
 int xquit;			/* exit if positive, force quit if negative */
 int xrow, xoff, xtop;		/* current row, column, and top row */
@@ -1290,6 +1291,7 @@ static void *eo_##opt(char *loc, char *cmd, char *arg) { inner }
 EO(pac) EO(pr) EO(ai) EO(ish) EO(ic) EO(grp) EO(shape) EO(seq)
 EO(sep) EO(tbs) EO(td) EO(order) EO(hll) EO(hlw) EO(hlp) EO(hlr)
 EO(hl) EO(lim) EO(led) EO(vis) EO(mpt) EO(err)
+EO(rec)
 
 _EO(left,
 	if (*loc)
@@ -1345,6 +1347,7 @@ static struct excmd {
 	{"m", ec_mark},
 	{"q!", ec_quit},
 	{"q", ec_quit},
+	EO(rec),
 	{"reg", ec_regprint},
 	{"rd", ec_undoredo},
 	{"r", ec_read},
diff --git a/term.c b/term.c
index 757c26eb..5655fd17 100644
--- a/term.c
+++ b/term.c
@@ -160,6 +160,12 @@ int term_read(void)
 		}
 		ibuf_cnt = 1;
 		ibuf_pos = 0;
+		if (xrec && *ibuf) {
+			char buf[2];
+			buf[0] = *ibuf;
+			buf[1] = '\0';
+			ex_regput(xrec, buf, 1);
+		}
 	}
 	if (icmd_pos < sizeof(icmd))
 		icmd[icmd_pos++] = ibuf[ibuf_pos];
diff --git a/vi.h b/vi.h
index a6f6348d..f7272c3c 100644
--- a/vi.h
+++ b/vi.h
@@ -430,6 +430,7 @@ extern int xsep;
 extern int xlim;
 extern int xseq;
 extern int xerr;
+extern int xrec;
 /* global variables */
 extern int xquit;
 extern int xrow, xoff, xtop;
