diff --git a/ex.c b/ex.c
index 5bf89e3b..91817f7d 100644
--- a/ex.c
+++ b/ex.c
@@ -22,6 +22,7 @@ int xpac;			/* print autocomplete options */
 int xkwdcnt;			/* number of search kwd changes */
 int xbufcur;			/* number of active buffers */
 int xgrec;			/* global vi/ex recursion depth */
+int xrec;			/* input recoding register */
 struct buf *bufs;		/* main buffers */
 struct buf tempbufs[2];		/* temporary buffers, for internal use */
 struct buf *ex_buf;		/* current buffer */
@@ -1098,6 +1099,7 @@ static int eo_##opt(char *loc, char *cmd, char *arg) { inner }
 EO(pac) EO(pr) EO(ai) EO(ish) EO(ic) EO(grp) EO(shape) EO(seq)
 EO(sep) EO(tbs) EO(td) EO(order) EO(hll) EO(hlw) EO(hlp) EO(hlr)
 EO(hl) EO(lim) EO(led) EO(vis) EO(mpt)
+EO(rec)
 
 _EO(left,
 	if (*loc)
@@ -1149,6 +1151,7 @@ static struct excmd {
 	{"k", ec_mark},
 	{"q!", ec_quit},
 	{"q", ec_quit},
+	EO(rec),
 	{"reg", ec_regprint},
 	{"rd", ec_undoredo},
 	{"r", ec_read},
diff --git a/term.c b/term.c
index 40e0cb4c..7fa1d2f4 100644
--- a/term.c
+++ b/term.c
@@ -160,6 +160,12 @@ int term_read(void)
 		}
 		ibuf_cnt = 1;
 		ibuf_pos = 0;
+		if (xrec && *ibuf) {
+			char buf[2];
+			buf[0] = *ibuf;
+			buf[1] = '\0';
+			vi_regputraw(xrec, buf, 0, 1);
+		}
 	}
 	if (icmd_pos < sizeof(icmd))
 		icmd[icmd_pos++] = ibuf[ibuf_pos];
diff --git a/vi.h b/vi.h
index ae224584..7359b8a1 100644
--- a/vi.h
+++ b/vi.h
@@ -500,6 +500,7 @@ extern int xpr;
 extern int xsep;
 extern int xlim;
 extern int xseq;
+extern int xrec;
 extern rset *xkwdrs;
 extern sbuf *xacreg;
 extern rset *fsincl;
