diff --git a/ex.c b/ex.c
index 734b51d..00ee6da 100644
--- a/ex.c
+++ b/ex.c
@@ -22,6 +22,7 @@ int xpac;			/* print autocomplete options */
 int xkwdcnt;			/* number of search kwd changes */
 int xbufcur;			/* number of active buffers */
 int xgrec;			/* global vi/ex recursion depth */
+int xrec;			/* input recoding register */
 struct buf *bufs;		/* main buffers */
 struct buf tempbufs[2];		/* temporary buffers, for internal use */
 struct buf *ex_buf;		/* current buffer */
@@ -913,6 +914,7 @@ static struct option {
 	{"sep", &xsep},
 	{"lim", &xlim},
 	{"seq", &xseq},
+	{"rec", &xrec},
 };
 
 static int ec_set(char *loc, char *cmd, char *arg)
diff --git a/term.c b/term.c
index aad65b8..b23b7f0 100644
--- a/term.c
+++ b/term.c
@@ -160,6 +160,12 @@ int term_read(void)
 		}
 		ibuf_cnt = 1;
 		ibuf_pos = 0;
+		if (xrec && *ibuf) {
+			char buf[2];
+			buf[0] = *ibuf;
+			buf[1] = '\0';
+			vi_regputraw(xrec, buf, 0, 1);
+		}
 	}
 	if (icmd_pos < sizeof(icmd))
 		icmd[icmd_pos++] = ibuf[ibuf_pos];
diff --git a/vi.h b/vi.h
index abd3ed6..32ac9f2 100644
--- a/vi.h
+++ b/vi.h
@@ -503,6 +503,7 @@ extern int xpr;
 extern int xsep;
 extern int xlim;
 extern int xseq;
+extern int xrec;
 extern rset *xkwdrs;
 extern sbuf *xacreg;
 extern rset *fsincl;
