diff --git a/ex.c b/ex.c
index 2419168..dd7943e 100644
--- a/ex.c
+++ b/ex.c
@@ -205,11 +205,12 @@ static char *ex_pathexpand(char *src)
 /* set the current search keyword rset if the kwd or flags changed */
 void ex_krsset(char *kwd, int dir)
 {
+	char *pat = {kwd};
 	char *reg = xregs['/'];
 	if (kwd && *kwd && ((!reg || !xkwdrs || strcmp(kwd, reg))
 		|| ((xkwdrs->regex->flg & REG_ICASE) != xic))) {
 		rset_free(xkwdrs);
-		xkwdrs = rset_make(1, (char*[]){kwd}, xic ? REG_ICASE : 0);
+		xkwdrs = rset_make(1, &pat, xic ? REG_ICASE : 0);
 		xkwdcnt++;
 		vi_regputraw('/', kwd, 0, 0);
 		xkwddir = dir;
@@ -809,7 +810,7 @@ static int ec_glob(char *loc, char *cmd, char *arg)
 	not = strchr(cmd, '!') || cmd[0] == 'v';
 	pat = re_read(&s);
 	if (pat)
-		rs = rset_make(1, (char*[]){pat}, xic ? REG_ICASE : 0);
+		rs = rset_make(1, &pat, xic ? REG_ICASE : 0);
 	free(pat);
 	if (!pat || !rs)
 		return 1;
@@ -958,9 +959,10 @@ static int ec_chdir(char *loc, char *cmd, char *arg)
 
 static int ec_setincl(char *loc, char *cmd, char *arg)
 {
+	char *pat = {arg};
 	rset_free(fsincl);
 	if (*arg)
-		fsincl = rset_make(1, (char*[]){arg}, xic ? REG_ICASE : 0);
+		fsincl = rset_make(1, &pat, xic ? REG_ICASE : 0);
 	else
 		fsincl = NULL;
 	return 0;
diff --git a/led.c b/led.c
index 9bef4a4..71a38fb 100644
--- a/led.c
+++ b/led.c
@@ -45,7 +45,8 @@ static void file_index(struct lbuf *buf)
 	char **ss = lbuf_buf(buf);
 	int ln_n = lbuf_len(buf), n;
 	sbuf *ibuf;
-	rset *rs = rset_make(1, (char*[]){xacreg ? xacreg->s : reg}, xic ? REG_ICASE : 0);
+	char *pat = {xacreg ? xacreg->s : reg};
+	rset *rs = rset_make(1, &pat, xic ? REG_ICASE : 0);
 	if (!rs)
 		return;
 	int subs[rs->grpcnt * 2];
@@ -287,10 +288,11 @@ static void led_printparts(sbuf *sb, int ps, char *post)
 		return;
 	int off, pos, psn = sb->s_n;
 	int idir = 0, next = sb->s[ps];
+	char **c = {NULL};
 	sbuf_str(sb, post)
 	sbuf_set(sb, '\0', 4)
 	rstate->s = NULL;
-	ren_position(sb->s+ps, &(char**){NULL}, &off);
+	ren_position(sb->s+ps, &c, &off);
 	off -= uc_slen(post);
 	pos = ren_cursor(sb->s+ps, ren_pos(sb->s+ps, MAX(0, off - 1)));
 	if (pos >= xleft + xcols)
diff --git a/uc.c b/uc.c
index 87d24c5..2d3695e 100644
--- a/uc.c
+++ b/uc.c
@@ -90,6 +90,13 @@ char *uc_subl(char *s, int beg, int end, int *rlen)
 	return r;
 }
 
+char *uc_sub(char *s, int beg, int end)
+{
+	int rlen = 0;
+
+	return uc_subl(s, beg, end, &rlen);
+}
+
 char *uc_dup(const char *s)
 {
 	char *r = emalloc(strlen(s) + 1);
diff --git a/vi.h b/vi.h
index 52eeab4..893b4b1 100644
--- a/vi.h
+++ b/vi.h
@@ -221,8 +221,8 @@ int uc_wid(int c);
 int uc_slen(char *s);
 char *uc_chr(char *s, int off);
 int uc_off(char *s, int off);
+char *uc_sub(char *s, int beg, int end);
 char *uc_subl(char *s, int beg, int end, int *rlen);
-#define uc_sub(s, beg, end) uc_subl(s, beg, end, &(int){0})
 char *uc_dup(const char *s);
 int uc_isspace(char *s);
 int uc_isprint(char *s);
