diff --git a/conf.c b/conf.c
index 926d306f..1f293a89 100644
--- a/conf.c
+++ b/conf.c
@@ -52,7 +52,7 @@ char msg_ft2[] = "/>>";	/* ex message (is never '\n' terminated) */
 #define CY1	14	/* bright cyan */
 #define WH1	15	/* bright white */
 
-#define A(...) (int[]){__VA_ARGS__}
+#define A(...) {__VA_ARGS__}
 
 /* At least 1 entry is required in this struct for fallback */
 /* lbuf lines are *always "\n\0" terminated, for $ to work one needs to account for '\n' too */
diff --git a/ex.c b/ex.c
index 2d75f3be..20ad8920 100644
--- a/ex.c
+++ b/ex.c
@@ -950,7 +950,7 @@ static void *ec_substitute(char *loc, char *cmd, char *arg)
 		rep = re_read(&s);
 	}
 	free(pat);
-	int offs[rs->nsubc];
+	int *offs = emalloc(rs->nsubc * sizeof(int));
 	for (i = beg; i < end; i++) {
 		char *ln = lbuf_get(xb, i);
 		sbuf *r = NULL;
@@ -1002,6 +1002,7 @@ static void *ec_substitute(char *loc, char *cmd, char *arg)
 	if (rs != xkwdrs)
 		rset_free(rs);
 	free(rep);
+	free(offs);
 	return NULL;
 }
 
diff --git a/lbuf.c b/lbuf.c
index 02ff38b1..fc65a009 100644
--- a/lbuf.c
+++ b/lbuf.c
@@ -437,7 +437,7 @@ int lbuf_search(struct lbuf *lb, rset *re, int dir, int beg, int end, int pskip,
 		int nskip, int *r, int *o)
 {
 	int r0 = *r, o0 = *o;
-	int offs[re->nsubc], i = r0;
+	int *offs = emalloc(re->nsubc * sizeof(int)), i = r0;
 	char *s = lbuf_get(lb, i);
 	int off, g1, g2, _o, step, flg;
 	if (pskip >= 0 && s)
@@ -462,14 +462,17 @@ int lbuf_search(struct lbuf *lb, rset *re, int dir, int beg, int end, int pskip,
 				break;
 			*o = _o;
 			*r = i;
-			if (dir > 0)
+			if (dir > 0) {
+				free(offs);
 				return 0;
+			}
 			step = off + g1;
 			off += g2 > 0 ? g2 : 1;
 			end = -1; /* break outer loop efficiently */
 		}
 		off = 0;
 	}
+	free(offs);
 	return end < 0 ? 0 : 1;
 }
 
diff --git a/led.c b/led.c
index b41abf0e..7d900391 100644
--- a/led.c
+++ b/led.c
@@ -47,7 +47,7 @@ static void file_index(struct lbuf *buf)
 		xic ? REG_ICASE | REG_NEWLINE : REG_NEWLINE);
 	if (!rs)
 		return;
-	int subs[rs->nsubc];
+	int *subs = emalloc(rs->nsubc * sizeof(int));
 	sbuf_smake(ibuf, 1024)
 	for (n = 1; n <= acsb->s_n; n++)
 		if (acsb->s[n - 1] == '\n')
@@ -80,6 +80,7 @@ static void file_index(struct lbuf *buf)
 	sbuf_null(acsb)
 	free(ibuf->s);
 	rset_free(rs);
+	free(subs);
 }
 
 static char *kmap_map(int kmap, int c)
@@ -150,10 +151,10 @@ void led_render(char *s0, int cbeg, int cend)
 	int att_old = 0, atti = 0, cterm = cend - cbeg;
 	char *bound = NULL;
 	char **chrs = r->chrs;	/* chrs[i]: the i-th character in s0 */
-	int off[cterm+1];	/* off[i]: the character at screen position i */
-	int att[cterm+1];	/* att[i]: the attributes of i-th character */
-	int stt[cterm+1];	/* stt[i]: remap off indexes */
-	int ctt[cterm+1];	/* ctt[i]: cterm bound attrs */
+	int *off = emalloc((cterm+1) * sizeof(int));	/* off[i]: the character at screen position i */
+	int *att = emalloc((cterm+1) * sizeof(int));	/* att[i]: the attributes of i-th character */
+	int *stt = NULL;	/* stt[i]: remap off indexes */
+	int *ctt = NULL;	/* ctt[i]: cterm bound attrs */
 	int ctx = r->ctx;
 	off[cterm] = -1;
 	if (ctx < 0) {
@@ -175,6 +176,8 @@ void led_render(char *s0, int cbeg, int cend)
 		if (o >= 0 && r->cmax > cterm && r->pos[o] + r->wid[o] > cend)
 			while (off[i] == o)
 				off[ctx < 0 ? i++ : i--] = -1;
+		stt = emalloc((cterm+1) * sizeof(int));
+		ctt = emalloc((cterm+1) * sizeof(int));
 		for (i = 0, c = 0; i < cterm;) {
 			if ((o = off[i++]) >= 0) {
 				att[c++] = o;
@@ -257,6 +260,10 @@ void led_render(char *s0, int cbeg, int cend)
 		memcpy(chrs[n], r->nullhole, r->holelen);
 		r->holelen = 0;
 	}
+	free(off);
+	free(att);
+	free(stt);
+	free(ctt);
 }
 
 static int led_lastchar(char *s)
diff --git a/regex.c b/regex.c
index 4a0addb8..1fc7adc5 100644
--- a/regex.c
+++ b/regex.c
@@ -399,7 +399,15 @@ static int reg_comp(rcode *prog, char *re, int nsubc, int laidx, int flg)
 	return 0;
 }
 
-#define _return(state) { if (eol_ch) utf8_length[eol_ch] = 1; return state; } \
+#define _return(state) { if (eol_ch) utf8_length[eol_ch] = 1; \
+free(pcs); \
+free(subs); \
+free(sdense); \
+free(_clist); \
+free(_nlist); \
+free(nsubs); \
+free(lb); \
+return state; } \
 
 #define newsub(init, copy) \
 if (freesub) { \
@@ -625,18 +633,19 @@ static int re_pikevm(rcode *prog, const char *s, const char **subp, int nsubc, i
 		return 0;
 	flg = prog->flg | flg;
 	const char *sp = s, *_sp = s, *s0, *s1;
-	int *pcs[prog->splits], *npc, *pc, *insts = prog->insts;
-	rsub *subs[prog->splits];
+	int *npc, *pc, *insts = prog->insts;
+	int **pcs = emalloc((prog->splits) * sizeof(int*));
+	rsub **subs = emalloc(prog->splits * sizeof(rsub*));
+	unsigned int *sdense = emalloc(prog->sparsesz * sizeof(unsigned int)), sparsesz = 0;
 	rsub *nsub, *sub, *matched = NULL, *freesub = NULL;
-	rthread _clist[prog->len], _nlist[prog->len];
+	rthread *_clist = emalloc(prog->len * sizeof(rthread)), *_nlist = emalloc(prog->len * sizeof(rthread));
 	rthread *clist = _clist, *nlist = _nlist, *tmp;
-	const char *_subp[2], *lb[prog->laidx+1];
+	const char *_subp[2], **lb = emalloc(prog->laidx * sizeof(lb[0]));
 	int rsubsize = prog->presub, suboff = 0;
 	int cnt, spc, i, c, j, osubp = nsubc * sizeof(char*);
 	int si = 0, clistidx = 0, nlistidx, mcont = MATCH;
 	int eol_ch = flg & REG_NEWLINE ? '\n' : 0;
-	unsigned int sdense[prog->sparsesz], sparsesz = 0;
-	char nsubs[prog->sub];
+	char *nsubs = emalloc(prog->sub);
 	for (i = 0; i < prog->laidx; i++)
 		lb[i] = NULL;
 	if (eol_ch)
@@ -714,7 +723,7 @@ rset *rset_make(int n, char **re, int flg)
 /* return the index of the matching regular expression or -1 if none matches */
 int rset_find(rset *rs, char *s, int *grps, int flg)
 {
-	const char *subs[rs->nsubc+2];
+	const char **subs = emalloc((rs->nsubc+2) * sizeof(subs[0]));
 	const char **sub = subs+2;
 	if (re_pikevm(rs->regex, s, sub, rs->nsubc, flg)) {
 		subs[1] = NULL; /* make sure sub[-1] never matches */
@@ -731,10 +740,12 @@ int rset_find(rset *rs, char *s, int *grps, int flg)
 						grps[gi + 1] = -1;
 					}
 				}
+				free(subs);
 				return i;
 			}
 		}
 	}
+	free(subs);
 	return -1;
 }
 
diff --git a/ren.c b/ren.c
index 8d68d054..0df9fc61 100644
--- a/ren.c
+++ b/ren.c
@@ -252,13 +252,14 @@ int syn_blockhl;
 static int syn_initft(int fti, int n, char *name, int flg)
 {
 	int i = n, set = hls[i].set;
-	char *pats[hlslen];
+	char **pats = emalloc(hlslen * sizeof(char *));
 	for (; i < hlslen && hls[i].ft == name && hls[i].set == set; i++)
 		pats[i - n] = hls[i].pat;
 	ftmap[fti].setbidx = n;
 	ftmap[fti].ft = name;
 	ftmap[fti].rs = rset_make(i - n, pats, flg);
 	ftmap[fti].seteidx = i;
+	free(pats);
 	return i < hlslen && hls[i].ft == name && hls[i].set != set;
 }
 
@@ -306,7 +307,7 @@ void syn_highlight(int *att, char *s, int n)
 	int fti = ftidx, blockhl = syn_blockhl, blockca = -1;
 	re:;
 	rset *rs = ftmap[fti].rs;
-	int subs[rs->nsubc], *catt, *iatt, sl, c;
+	int *subs = emalloc(rs->nsubc * sizeof(int)), *catt, *iatt, sl, c;
 	int cend, sidx = 0, flg = 0, hl, j, i, ii;
 	while ((sl = rset_find(rs, s + sidx, subs, flg)) >= 0) {
 		cend = 1;
@@ -349,15 +350,20 @@ void syn_highlight(int *att, char *s, int n)
 		flg = REG_NOTBOL;
 	}
 	fti++;
-	if (ftmidx > fti && ftmap[fti-1].ft == ftmap[fti].ft)
+	if (ftmidx > fti && ftmap[fti-1].ft == ftmap[fti].ft) {
+		free(subs);
 		goto re;
+	}
 	if (blockca >= 0 && SYN_BSSET(blockca) && !SYN_BESET(blockca) && last_scdir > 0)
 		syn_blockhl = -1;
-	if (syn_blockhl < 0 || blockhl < 0)
+	if (syn_blockhl < 0 || blockhl < 0) {
+		free(subs);
 		return;
+	}
 	for (j = 0; j < n; j++)
 		if (!att[j] || !SYN_BPSET(blockatt))
 			att[j] = blockatt;
+	free(subs);
 }
 
 char *syn_filetype(char *path)
@@ -401,9 +407,10 @@ int syn_addhl(char *reg, int id)
 
 void syn_init(void)
 {
-	char *pats[ftslen];
+	char **pats = emalloc(ftslen * sizeof(char *));
 	int i = 0;
 	for (; i < ftslen; i++)
 		pats[i] = fts[i].pat;
 	syn_ftrs = rset_make(i, pats, 0);
+	free(pats);
 }
diff --git a/vi.c b/vi.c
index b78c070b..cb7e804d 100644
--- a/vi.c
+++ b/vi.c
@@ -140,7 +140,8 @@ static void vi_drawrow(int row)
 			vi_rshift = 0;
 		if (row != xrow+1 || !c || *c == '\n')
 			goto skip;
-		char tmp[xcols+3], snum[32];
+		char *tmp = emalloc(xcols+3);
+		char *snum = emalloc(32);
 		memset(tmp, ' ', xcols+1);
 		tmp[xcols+1] = '\n';
 		tmp[xcols+2] = '\0';
@@ -166,6 +167,8 @@ static void vi_drawrow(int row)
 		restore(xtd)
 		restore(ftidx)
 		vi_rshift = (row != xtop + xrows-1);
+		free(tmp);
+		free(snum);
 		return;
 	}
 	s = lbuf_get(xb, row);
@@ -1287,10 +1290,11 @@ void vi(int init)
 				if (!(ln = lbuf_get(xb, xrow)))
 					break;
 				ln += xoff;
-				char buf[strlen(ln)+4];
+				char *buf = emalloc(strlen(ln)+4);
 				strcpy(buf, ":e ");
 				strcpy(buf+3, ln);
 				term_push(buf, strlen(ln)+3);
+				free(buf);
 				break; }
 			case TK_CTL('n'):
 				vi_cndir = vi_arg ? -vi_cndir : vi_cndir;
@@ -1410,7 +1414,7 @@ void vi(int init)
 				case 't': {
 					vi_drawmsg("arg2:(0|#)");
 					cs = vi_curword(xb, xrow, xoff, vi_prefix());
-					char buf[cs ? strlen(cs)+30 : 30];
+					char *buf = emalloc(cs ? strlen(cs)+30 : 30);
 					strcpy(buf, ".,.+");
 					char *buf1 = itoa(vi_arg, buf+4);
 					strcat(buf1, "s/");
@@ -1420,10 +1424,11 @@ void vi(int init)
 						free(cs);
 					}
 					ln = vi_enprompt(":", buf, &k, &n);
+					free(buf);
 					goto do_excmd; }
 				case 'r': {
 					cs = vi_curword(xb, xrow, xoff, vi_arg);
-					char buf[cs ? strlen(cs)+30 : 30];
+					char *buf = emalloc(cs ? strlen(cs)+30 : 30);
 					strcpy(buf, "%s/");
 					if (cs) {
 						strcat(buf, cs);
@@ -1431,6 +1436,7 @@ void vi(int init)
 						free(cs);
 					}
 					ln = vi_enprompt(":", buf, &k, &n);
+					free(buf);
 					goto do_excmd; }
 				default:
 					term_dec()
diff --git a/vi.h b/vi.h
index 0c3e52bb..34c1bd61 100644
--- a/vi.h
+++ b/vi.h
@@ -494,7 +494,7 @@ extern const int ftslen;
 struct highlight {
 	char *ft;		/* the filetype of this pattern */
 	char *pat;		/* regular expression */
-	int *att;		/* attributes of the matched groups */
+	int att[16];		/* attributes of the matched groups */
 	unsigned char set;	/* subset index */
 	unsigned char id;	/* id of this hl */
 };
