diff --git a/ex.c b/ex.c
index e41edeb..d08b293 100644
--- a/ex.c
+++ b/ex.c
@@ -28,7 +28,7 @@ struct buf *ex_buf;		/* current buffer */
 struct buf *ex_pbuf;		/* prev buffer */
 static struct buf *ex_tpbuf;	/* temp prev buffer */
 sbuf *xacreg;			/* autocomplete db filter regex */
-rset *xkwdrs;			/* the last searched keyword rset */
+rstr *xkwdrs;			/* the last searched keyword rstr */
 int xkwddir;			/* the last search direction */
 int xmpt;			/* whether to prompt after printing > 1 lines in vi */
 int xpr;			/* ex_cprint register */
@@ -200,9 +200,9 @@ void ex_krsset(char *kwd, int dir)
 {
 	char *reg = xregs['/'];
 	if (kwd && *kwd && ((!reg || !xkwdrs || strcmp(kwd, reg))
-		|| ((xkwdrs->regex->flg & REG_ICASE) != xic))) {
-		rset_free(xkwdrs);
-		xkwdrs = rset_smake(kwd, xic ? REG_ICASE : 0);
+		|| (xkwdrs->icase != xic))) {
+		rstr_free(xkwdrs);
+		xkwdrs = rstr_make(kwd, xic ? REG_ICASE : 0);
 		xkwdcnt++;
 		vi_regputraw('/', kwd, 0, 0);
 		xkwddir = dir;
@@ -772,11 +772,11 @@ static int ec_substitute(char *loc, char *cmd, char *arg)
 		ex_print("re syntax error");
 		return 1;
 	}
-	int offs[xkwdrs->grpcnt * 2];
+	int offs[xkwdrs->rs ? xkwdrs->rs->grpcnt * 2 : 2];
 	for (i = beg; i < end; i++) {
 		char *ln = lbuf_get(xb, i);
 		sbuf *r = NULL;
-		while (rset_find(xkwdrs, ln, offs, REG_NEWLINE) >= 0) {
+		while (rstr_find(xkwdrs, ln, offs, REG_NEWLINE) >= 0) {
 			if (offs[xgrp] < 0) {
 				ln += offs[1] > 0 ? offs[1] : 1;
 				continue;
@@ -851,7 +851,7 @@ static int ec_glob(char *loc, char *cmd, char *arg)
 	int beg, end, not;
 	char *pat, *s = arg;
 	int i;
-	rset *rs;
+	rstr *rs;
 	if (!loc[0] && !xgdep)
 		loc = "%";
 	if (ex_region(loc, &beg, &end))
@@ -859,7 +859,7 @@ static int ec_glob(char *loc, char *cmd, char *arg)
 	not = !!strchr(cmd, '!');
 	pat = re_read(&s);
 	if (pat)
-		rs = rset_smake(pat, xic ? REG_ICASE : 0);
+		rs = rstr_make(pat, xic ? REG_ICASE : 0);
 	free(pat);
 	if (!pat || !rs)
 		return 1;
@@ -869,7 +869,7 @@ static int ec_glob(char *loc, char *cmd, char *arg)
 	for (i = beg; i < lbuf_len(xb);) {
 		char *ln = lbuf_get(xb, i);
 		lbuf_s(ln)->grec &= ~xgdep;
-		if ((rset_find(rs, ln, NULL, REG_NEWLINE) < 0) == not) {
+		if ((rstr_find(rs, ln, NULL, REG_NEWLINE) < 0) == not) {
 			xrow = i;
 			if (ex_exec(s))
 				break;
@@ -878,7 +878,7 @@ static int ec_glob(char *loc, char *cmd, char *arg)
 		while (i < lbuf_len(xb) && !(lbuf_i(xb, i)->grec & xgdep))
 			i++;
 	}
-	rset_free(rs);
+	rstr_free(rs);
 	xgdep /= 2;
 	return 0;
 }
@@ -1009,10 +1009,10 @@ static int ec_chdir(char *loc, char *cmd, char *arg)
 
 static int ec_setincl(char *loc, char *cmd, char *arg)
 {
-	rset_free(fsincl);
+	rstr_free(fsincl);
 	if (!*arg)
 		fsincl = NULL;
-	else if (!(fsincl = rset_smake(arg, xic ? REG_ICASE : 0)))
+	else if (!(fsincl = rstr_make(arg, xic ? REG_ICASE : 0)))
 		ex_print("syntax error");
 	return 0;
 }
diff --git a/lbuf.c b/lbuf.c
index 0649f71..cea68ac 100644
--- a/lbuf.c
+++ b/lbuf.c
@@ -359,16 +359,16 @@ int lbuf_findchar(struct lbuf *lb, char *cs, int cmd, int n, int *row, int *off)
 	return n != 0;
 }
 
-int lbuf_search(struct lbuf *lb, rset *re, int dir, int *r,
+int lbuf_search(struct lbuf *lb, rstr *re, int dir, int *r,
 			int *o, int ln_n, int skip)
 {
 	int r0 = *r, o0 = *o;
-	int offs[re->grpcnt * 2], i = r0;
+	int offs[re->rs ? re->rs->grpcnt * 2 : 2], i = r0;
 	char *s = lbuf_get(lb, i);
 	int off = skip >= 0 && *uc_chr(s, o0 + skip) ? uc_chr(s, o0 + skip) - s : 0;
 	for (; i >= 0 && i < ln_n; i += dir) {
 		s = lb->ln[i];
-		while (rset_find(re, s + off, offs,
+		while (rstr_find(re, s + off, offs,
 				off ? REG_NOTBOL | REG_NEWLINE : REG_NEWLINE) >= 0) {
 			int g1 = offs[xgrp], g2 = offs[xgrp + 1];
 			if (g1 < 0) {
diff --git a/regex.c b/regex.c
index 08470a5..de613fb 100644
--- a/regex.c
+++ b/regex.c
@@ -713,3 +713,126 @@ char *re_read(char **src)
 	*src = *s ? s + 1 : s;
 	sbufn_sret(sb)
 }
+
+/* return zero if a simple pattern is given */
+static int rstr_simple(rstr *rs, char *re, int icase)
+{
+	char *beg;
+	char *end;
+	if (!strcmp(re, "^$"))
+		return 1;
+	rs->lbeg = re[0] == '^';
+	if (rs->lbeg)
+		re++;
+	rs->wbeg = re[0] == '\\' && re[1] == '<';
+	if (rs->wbeg)
+		re += 2;
+	beg = re;
+	while (re[0] && !strchr("\\.*+?[]{}()$", (unsigned char) re[0]))
+		re++;
+	end = re;
+	rs->wend = re[0] == '\\' && re[1] == '>';
+	if (rs->wend)
+		re += 2;
+	rs->lend = re[0] == '$';
+	if (rs->lend)
+		re++;
+	if (!re[0]) {
+		int len = end - beg;
+		rs->len = len;
+		rs->str = emalloc(len + 1);
+		rs->str[len] = '\0';
+		if (icase) {
+			while (--len >= 0)
+				rs->str[len] = tolower((unsigned char)beg[len]);
+		} else
+			memcpy(rs->str, beg, len);
+		return 0;
+	}
+	return 1;
+}
+
+rstr *rstr_make(char *re, int flg)
+{
+	rstr *rs = emalloc(sizeof(*rs));
+	memset(rs, 0, sizeof(*rs));
+	rs->icase = flg & REG_ICASE;
+	if (rstr_simple(rs, re, rs->icase))
+		rs->rs = rset_make(1, &re, flg);
+	if (!rs->rs && !rs->str) {
+		free(rs);
+		return NULL;
+	}
+	return rs;
+}
+
+#define rstr_match(gen, wbeg, wend, cmpcase) \
+wbeg wend \
+m = rs->str; t = r; \
+for (; *m && *t; t++, m++) { \
+	if (cmpcase) \
+		goto break##gen; \
+} \
+if (!*m) { \
+	if (grps) { \
+		grps[0] = r - s; \
+		grps[1] = r - s + len; \
+	} \
+	return 0; \
+} \
+break##gen:; \
+
+#define rstr_match1(gen, wbeg, wend, cmpcase) \
+{ for (r = beg; r <= end; r++) { \
+	rstr_match(2##gen, wbeg, wend, cmpcase) \
+} } \
+
+#define _wbeg if (r > s && (isword(r - 1) || !isword(r))) continue;
+#define _wend if (r[len] && (!isword(r + len - 1) || isword(r + len))) continue;
+
+#define template(gen, cmpcase) \
+if (!rs->wbeg && !rs->wend) \
+	rstr_match1(1##gen, /*nop*/, /*nop*/, cmpcase) \
+else if (rs->wbeg && !rs->wend) \
+	rstr_match1(2##gen, _wbeg, /*nop*/, cmpcase) \
+else if (!rs->wbeg && rs->wend) \
+	rstr_match1(3##gen, /*nop*/, _wend, cmpcase) \
+else \
+	rstr_match1(4##gen, _wbeg, _wend, cmpcase) \
+
+/* return zero if an occurrence is found */
+int rstr_find(rstr *rs, char *s, int *grps, int flg)
+{
+	int len;
+	char *beg, *end, *r, *t, *m;
+	if (rs->rs)
+		return rset_find(rs->rs, s, grps, flg);
+	if ((rs->lbeg && (flg & REG_NOTBOL)) || (rs->lend && (flg & REG_NOTEOL)))
+		return -1;
+	if (!*s)
+		return -1;
+	len = rs->len;
+	beg = s;
+	end = s + strlen(s) - len - (flg & REG_NEWLINE ? 1 : 0);
+	if (end < beg)
+		return -1;
+	if (rs->lend)
+		beg = end;
+	if (rs->lbeg)
+		end = s;
+	if (!rs->icase) {
+		template(3, *t != *m)
+	} else {
+		template(4, tolower((unsigned char) *t) != *m)
+	}
+	return -1;
+}
+
+void rstr_free(rstr *rs)
+{
+	if (!rs)
+		return;
+	rset_free(rs->rs);
+	free(rs->str);
+	free(rs);
+}
diff --git a/vi.c b/vi.c
index 3ca8bc4..5d1fff5 100644
--- a/vi.c
+++ b/vi.c
@@ -525,7 +525,7 @@ void vi_regput(int c, const char *s, int ln)
 	vi_regputraw(c, s, ln, isupper(c));
 }
 
-rset *fsincl;
+rstr *fsincl;
 char *fs_exdir;
 static int fspos;
 static int fsdir;
@@ -565,7 +565,7 @@ void dir_calc(char *path)
 				memcpy(cpath, ptrs[i], pathlen + len);
 				plen[i++] = pathlen + len;
 			} else if (ret >= 0 && S_ISREG(statbuf.st_mode))
-				if (!fsincl || rset_find(fsincl, cpath, NULL, 0) >= 0) {
+				if (!fsincl || rstr_find(fsincl, cpath, NULL, 0) >= 0) {
 					sbuf_mem(sb, cpath, (int)(pathlen + len))
 					sbuf_chr(sb, '\n')
 				}
diff --git a/vi.h b/vi.h
index f00adb9..a3b9221 100644
--- a/vi.h
+++ b/vi.h
@@ -122,12 +122,23 @@ typedef struct {
 	int *setgrpcnt;		/* number of groups in each regular expression */
 	int grpcnt;		/* group count */
 } rset;
+typedef struct {
+	rset *rs;		/* only for regex patterns */
+	char *str;		/* for simple, non-regex patterns  */
+	int len;		/* str length */
+	int icase;		/* ignore case */
+	int lbeg, lend;		/* match line beg/end */
+	int wbeg, wend;		/* match word beg/end */
+} rstr;
 rset *rset_make(int n, char **pat, int flg);
 rset *rset_smake(char *pat, int flg)
 	{ char *ss[1] = {pat}; return rset_make(1, ss, flg); }
 int rset_find(rset *re, char *s, int *grps, int flg);
 void rset_free(rset *re);
 char *re_read(char **src);
+rstr *rstr_make(char *re, int flg);
+int rstr_find(rstr *rs, char *s, int *grps, int flg);
+void rstr_free(rstr *rs);
 
 /* lbuf.c line buffer, managing a number of lines */
 #define NMARKS_BASE		('z' - 'a' + 2)
@@ -180,7 +191,7 @@ void lbuf_saved(struct lbuf *lb, int clear);
 int lbuf_indents(struct lbuf *lb, int r);
 int lbuf_eol(struct lbuf *lb, int r);
 int lbuf_findchar(struct lbuf *lb, char *cs, int cmd, int n, int *r, int *o);
-int lbuf_search(struct lbuf *lb, rset *re, int dir, int *r,
+int lbuf_search(struct lbuf *lb, rstr *re, int dir, int *r,
 			int *o, int ln_n, int skip);
 #define lbuf_dedup(lb, str, n) \
 { for (int i = 0; i < lbuf_len(lb);) { \
@@ -401,7 +412,7 @@ void ex_cprint(char *line, int r, int c, int ln);
 #define ex_print(line) ex_cprint(line, -1, 0, 1)
 void ex_init(char **files, int n);
 void ex_bufpostfix(struct buf *p, int clear);
-int ex_krs(rset **krs, int *dir);
+int ex_krs(rstr **krs, int *dir);
 void ex_krsset(char *kwd, int dir);
 int ex_edit(const char *path, int len);
 void ec_bufferi(int id);
@@ -499,9 +510,9 @@ extern int xkwddir;
 extern int xmpt;
 extern int xpr;
 extern int xsep;
-extern rset *xkwdrs;
+extern rstr *xkwdrs;
 extern sbuf *xacreg;
-extern rset *fsincl;
+extern rstr *fsincl;
 extern char *fs_exdir;
 extern int vi_hidch;
 extern int vi_insmov;
