diff --git a/ex.c b/ex.c
index c5b722d..f9de211 100644
--- a/ex.c
+++ b/ex.c
@@ -381,6 +381,8 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 			return 1;
 	if (!(path = ex_pathexpand(arg, 0)))
 		return 1;
+	if (!loc)
+		goto ret;
 	if (path[0] == '.' && path[1] == '/')
 		cd = 2;
 	if (path[0] && ((fd = bufs_find(path+cd)) >= 0)) {
@@ -393,6 +395,7 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 	if (!rd && ex_buf - bufs < xbufcur && ex_buf - bufs >= 0) {
 		ex_bufpostfix(ex_buf - bufs, path[0]);
 		syn_setft(ex_buf->ft);
+		ret:
 		snprintf(msg, sizeof(msg), "\"%s\"  %d lines  [r]",
 				*ex_path ? ex_path : "unnamed", lbuf_len(xb));
 		ex_show(msg);
@@ -1114,5 +1117,9 @@ int ex_init(char **files)
 		return 1;
 	if (xled && (s = getenv("EXINIT")))
 		ex_command(s)
+	if (stdin_fd) {
+		lbuf_rd(xb, STDIN_FILENO, 0, lbuf_len(xb));
+		ec_edit(NULL, "e!", "");
+	}
 	return 0;
 }
diff --git a/term.c b/term.c
index 2716c2d..6578f77 100644
--- a/term.c
+++ b/term.c
@@ -1,5 +1,6 @@
 sbuf *term_sbuf;
 int term_record;
+int stdin_fd;
 static int rows, cols;
 static struct termios termios;
 
@@ -10,16 +11,16 @@ void term_init(void)
 	struct winsize win;
 	struct termios newtermios;
 	sbufn_make(term_sbuf, 2048)
-	tcgetattr(0, &termios);
+	tcgetattr(stdin_fd, &termios);
 	newtermios = termios;
 	newtermios.c_lflag &= ~(ICANON | ISIG);
 	newtermios.c_lflag &= ~ECHO;
-	tcsetattr(0, TCSAFLUSH, &newtermios);
+	tcsetattr(stdin_fd, TCSAFLUSH, &newtermios);
 	if (getenv("LINES"))
 		rows = atoi(getenv("LINES"));
 	if (getenv("COLUMNS"))
 		cols = atoi(getenv("COLUMNS"));
-	if (!ioctl(0, TIOCGWINSZ, &win)) {
+	if (!ioctl(stdin_fd, TIOCGWINSZ, &win)) {
 		cols = win.ws_col;
 		rows = win.ws_row;
 	}
@@ -137,12 +138,12 @@ int term_read(void)
 	struct pollfd ufds[1];
 	int n;
 	if (ibuf_pos >= ibuf_cnt) {
-		ufds[0].fd = STDIN_FILENO;
+		ufds[0].fd = stdin_fd;
 		ufds[0].events = POLLIN;
 		if (poll(ufds, 1, -1) <= 0)
 			return -1;
 		/* read a single input character */
-		if ((n = read(STDIN_FILENO, ibuf, 1)) <= 0)
+		if ((n = read(stdin_fd, ibuf, 1)) <= 0)
 			return -1;
 		ibuf_cnt = n;
 		ibuf_pos = 0;
@@ -318,7 +319,7 @@ char *cmd_pipe(char *cmd, char *ibuf, int iproc, int oproc)
 	close(fds[1].fd);
 	waitpid(pid, NULL, 0);
 	signal(SIGTTOU, SIG_IGN);
-	tcsetpgrp(STDIN_FILENO, getpgrp());
+	tcsetpgrp(stdin_fd, getpgrp());
 	signal(SIGTTOU, SIG_DFL);
 	if (!iproc) {
 		term_init();
diff --git a/vi.c b/vi.c
index 64fc9a7..4bc75ac 100644
--- a/vi.c
+++ b/vi.c
@@ -2073,10 +2073,12 @@ int main(int argc, char *argv[])
 	for (i = 1; i < argc && argv[i][0] == '-'; i++) {
 		if (argv[i][1] == 's')
 			xled = 0;
-		if (argv[i][1] == 'e')
+		else if (argv[i][1] == 'e')
 			xvis = 0;
-		if (argv[i][1] == 'v')
+		else if (argv[i][1] == 'v')
 			xvis = 1;
+		else
+			stdin_fd = open(ctermid(NULL), O_RDONLY);
 	}
 	term_init();
 	if (!ex_init(argv + i)) {
diff --git a/vi.h b/vi.h
index c1485c1..755bd4b 100644
--- a/vi.h
+++ b/vi.h
@@ -426,3 +426,4 @@ extern int fspos;
 extern int fscount;
 extern int vi_hidch;
 extern int vi_insmov;
+extern int stdin_fd;
