diff --git a/ex.c b/ex.c
index 57b642e..96bc2e3 100644
--- a/ex.c
+++ b/ex.c
@@ -381,6 +381,8 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 {
 	char msg[128];
 	int fd, rd = 0, cd = 0;
+	if (!loc)
+		goto ret;
 	if (arg[0] == '.' && arg[1] == '/')
 		cd = 2;
 	if (arg[0] && ((fd = bufs_find(arg+cd)) >= 0)) {
@@ -396,6 +398,7 @@ static int ec_edit(char *loc, char *cmd, char *arg)
 	if (ex_buf - bufs < xbufcur && ex_buf - bufs >= 0) {
 		ex_bufpostfix(ex_buf - bufs, arg[0]);
 		syn_setft(ex_ft);
+		ret:
 		snprintf(msg, sizeof(msg), "\"%s\"  %d lines  [r]",
 				*ex_path ? ex_path : "unnamed", lbuf_len(xb));
 		ex_show(rd ? "read failed" : msg);
@@ -1075,4 +1078,8 @@ void ex_init(char **files, int n)
 	} while (--n > 0);
 	if (xled && (s = getenv("EXINIT")))
 		ex_command(s)
+	if (stdin_fd) {
+		lbuf_rd(xb, STDIN_FILENO, 0, lbuf_len(xb));
+		ec_edit(NULL, "e!", "");
+	}
 }
diff --git a/term.c b/term.c
index be3dbab..b131396 100644
--- a/term.c
+++ b/term.c
@@ -1,6 +1,7 @@
 sbuf *term_sbuf;
 int term_record;
 int xrows, xcols;
+int stdin_fd;
 static struct termios termios;
 
 void term_init(void)
@@ -10,16 +11,16 @@ void term_init(void)
 	struct winsize win;
 	struct termios newtermios;
 	sbufn_make(term_sbuf, 2048)
-	tcgetattr(0, &termios);
+	tcgetattr(stdin_fd, &termios);
 	newtermios = termios;
 	newtermios.c_lflag &= ~(ICANON | ISIG);
 	newtermios.c_lflag &= ~ECHO;
-	tcsetattr(0, TCSAFLUSH, &newtermios);
+	tcsetattr(stdin_fd, TCSAFLUSH, &newtermios);
 	if (getenv("LINES"))
 		xrows = atoi(getenv("LINES"));
 	if (getenv("COLUMNS"))
 		xcols = atoi(getenv("COLUMNS"));
-	if (!ioctl(0, TIOCGWINSZ, &win)) {
+	if (!ioctl(stdin_fd, TIOCGWINSZ, &win)) {
 		xcols = win.ws_col;
 		xrows = win.ws_row;
 	}
@@ -34,7 +35,7 @@ void term_done(void)
 		return;
 	term_commit();
 	sbuf_free(term_sbuf)
-	tcsetattr(0, 0, &termios);
+	tcsetattr(stdin_fd, 0, &termios);
 }
 
 void term_clean(void)
@@ -127,12 +128,12 @@ int term_read(void)
 	struct pollfd ufds[1];
 	int n;
 	if (ibuf_pos >= ibuf_cnt) {
-		ufds[0].fd = STDIN_FILENO;
+		ufds[0].fd = stdin_fd;
 		ufds[0].events = POLLIN;
 		if (poll(ufds, 1, -1) <= 0)
 			return -1;
 		/* read a single input character */
-		if ((n = read(STDIN_FILENO, ibuf, 1)) <= 0)
+		if ((n = read(stdin_fd, ibuf, 1)) <= 0)
 			return -1;
 		ibuf_cnt = n;
 		ibuf_pos = 0;
@@ -307,7 +308,7 @@ char *cmd_pipe(char *cmd, char *ibuf, int oproc)
 		close(ifd);
 	waitpid(pid, NULL, 0);
 	signal(SIGTTOU, SIG_IGN);
-	tcsetpgrp(STDIN_FILENO, getpgrp());
+	tcsetpgrp(stdin_fd, getpgrp());
 	signal(SIGTTOU, SIG_DFL);
 	if (!ibuf) {
 		term_init();
diff --git a/vi.c b/vi.c
index fa061ac..95ceec9 100644
--- a/vi.c
+++ b/vi.c
@@ -2060,10 +2060,12 @@ int main(int argc, char *argv[])
 	for (i = 1; i < argc && argv[i][0] == '-'; i++) {
 		if (argv[i][1] == 's')
 			xled = 0;
-		if (argv[i][1] == 'e')
+		else if (argv[i][1] == 'e')
 			xvis = 0;
-		if (argv[i][1] == 'v')
+		else if (argv[i][1] == 'v')
 			xvis = 1;
+		else
+			stdin_fd = open(ctermid(NULL), O_RDONLY);
 	}
 	term_init();
 	ex_init(argv + i, argc - i);
diff --git a/vi.h b/vi.h
index 4222f50..57f0964 100644
--- a/vi.h
+++ b/vi.h
@@ -424,3 +424,4 @@ extern int fspos;
 extern int fscount;
 extern int vi_hidch;
 extern int vi_insmov;
+extern int stdin_fd;
