diff --git a/lbuf.c b/lbuf.c
index 69db198..8b04ecf 100644
--- a/lbuf.c
+++ b/lbuf.c
@@ -255,6 +255,41 @@ int lbuf_jump(struct lbuf *lbuf, int mark, int *pos, int *off)
 	return 0;
 }
 
+int lbuf_undojump(struct lbuf *lb, int *pos, int *off)
+{
+	struct lopt *lo;
+	static int last_hist_u;
+	int hist_u = lb->hist_u;
+	int useq;
+	int rpos = 0;
+	int ret = 0;
+	if (!last_hist_u) {
+		last_hist_u = hist_u;
+		ret = 2;
+	}
+	if (hist_u) {
+		do {
+			lo = &lb->hist[hist_u - 1];
+			useq = lo->seq;
+			while (hist_u && lb->hist[hist_u - 1].seq == useq)
+				lo = &lb->hist[--hist_u];
+		} while (hist_u && hist_u >= last_hist_u);
+		last_hist_u = hist_u;
+		hist_u = lb->hist_u;
+		while (hist_u != last_hist_u) {
+			lo = &lb->hist[--hist_u];
+			if (lb->hist[last_hist_u].pos > lo->pos) {
+				rpos += lo->n_ins;
+				rpos -= lbuf_linecount(lo->del);
+			}
+		}
+	} else
+		return ret ? ret : 1;
+	*pos = rpos + lb->hist[last_hist_u].pos;
+	*off = lo->pos_off;
+	return ret;
+}
+
 int lbuf_undo(struct lbuf *lb)
 {
 	if (!lb->hist_u)
diff --git a/vi.c b/vi.c
index 3c09077..aeb943d 100644
--- a/vi.c
+++ b/vi.c
@@ -1568,6 +1568,17 @@ void vi(int init)
 				vi_hidch = !vi_hidch;
 				vi_mod |= 1;
 				break;
+			case TK_CTL('o'):
+				next_hop:
+				if (lbuf_undojump(xb, &xrow, &xoff))
+					snprintf(vi_msg, sizeof(vi_msg), "undo jmp failed");
+				else if (xrow == nrow)
+					goto next_hop;
+				vi_col = vi_off2col(xb, xrow, xoff);
+				xoff = vi_col2off(xb, xrow, vi_col);
+				xtop = MAX(0, xrow - xrows / 2);
+				vi_mod = 1;
+				break;
 			case TK_CTL('v'):
 				vi_arg1 = (vi_wsel % 5) + !!*vi_word;
 			case TK_CTL('c'):
diff --git a/vi.h b/vi.h
index 9e17595..52906d7 100644
--- a/vi.h
+++ b/vi.h
@@ -169,6 +169,7 @@ void lbuf_emark(struct lbuf *lb, int hist_n, int beg, int end);
 int lbuf_opt(struct lbuf *lb, char *buf, int pos, int n_del);
 void lbuf_mark(struct lbuf *lbuf, int mark, int pos, int off);
 int lbuf_jump(struct lbuf *lbuf, int mark, int *pos, int *off);
+int lbuf_undojump(struct lbuf *lb, int *pos, int *off);
 int lbuf_undo(struct lbuf *lbuf);
 int lbuf_redo(struct lbuf *lbuf);
 int lbuf_modified(struct lbuf *lb);
